<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        
        
        <link rel="shortcut icon" href="./img/favicon.ico">

        <title>My Docs</title>

        <link href="./css/bootstrap-custom.min.css" rel="stylesheet">
        <link href="./css/font-awesome-4.0.3.css" rel="stylesheet">
        <link href="./css/base.css" rel="stylesheet">
        <link rel="stylesheet" href="./css/highlight.css">

        <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
        <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

        
    </head>

    <body class="homepage">

        <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->
            <a class="navbar-brand" href=".">My Docs</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="original.index/">Original.index</a>
                    </li>
                
                
                </ul>
            

            
            <!-- Search, Navigation and Repo links -->
            <ul class="nav navbar-nav navbar-right">
                
                
                <li class="disabled">
                    <a rel="next" >
                        <i class="fa fa-arrow-left"></i> Previous
                    </a>
                </li>
                <li >
                    <a rel="prev" href="original.index/">
                        Next <i class="fa fa-arrow-right"></i>
                    </a>
                </li>
                
                
            </ul>
            
        </div>
    </div>
</div>

        <div class="container">
            
                <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="main active"><a href="#ngseasy-beta">NGSeasy (beta)</a></li>
        
            <li><a href="#a-dockerized-ngs-pipeline-and-tool-box">[A Dockerized NGS pipeline and tool-box]</a></li>
        
    
        <li class="main "><a href="#ngseasy-easy-analysis-of-next-generation-sequencing">NGSeasy (Easy Analysis of Next Generation Sequencing)</a></li>
        
            <li><a href="#advantages">Advantages</a></li>
        
    
        <li class="main "><a href="#overview-of-the-ngseasy-pipeline-components">Overview of the NGSeasy Pipeline Components</a></li>
        
    
        <li class="main "><a href="#the-full-ngseasy-pipeline">The Full NGSeasy pipeline</a></li>
        
            <li><a href="#coming-soon">Coming Soon</a></li>
        
    
        <li class="main "><a href="#dockerised-ngseasy">Dockerised NGSeasy</a></li>
        
            <li><a href="#the-12-easy-steps-to-ngs-freedom">The 12 Easy Steps to NGS Freedom!</a></li>
        
            <li><a href="#1-install-docker">1. Install Docker</a></li>
        
            <li><a href="#2-get-ngseasy-automated-build-container-images">2. Get NGSeasy Automated build Container Images</a></li>
        
            <li><a href="#3-make-ngseasy-project-directory">3. Make NGSeasy project directory</a></li>
        
            <li><a href="#4-download-ngseasy-resources">4. Download NGSeasy Resources</a></li>
        
            <li><a href="#5-get-ngseasy-sripts">5. Get NGSeasy Sripts</a></li>
        
            <li><a href="#6-manually-build-required-ngseasy-container-images">6. Manually Build required NGSeasy Container Images</a></li>
        
            <li><a href="#7-manually-build-ngseasy-variant-annotaion-container-images">7. Manually Build NGSeasy Variant Annotaion Container Images</a></li>
        
            <li><a href="#8-set-up-ngseasy-project-working-directories">8. Set up NGSeasy Project Working Directories</a></li>
        
            <li><a href="#ngs-project-directory">NGS Project Directory</a></li>
        
            <li><a href="#9-ngseasy-project-configuration-file">9. NGSeasy Project configuration file</a></li>
        
            <li><a href="#10-copy-project-fastq-files-to-relevent-projectsample-directories">10. Copy Project Fastq files to relevent Project/Sample Directories</a></li>
        
            <li><a href="#11-start-the-ngseasy-volume-contaier">11. Start the NGSeasy Volume Contaier</a></li>
        
            <li><a href="#12-running-an-ngseasy-full-pipeline-from-raw-fastq-to-vcf-calls">12. Running an NGSeasy full pipeline : from raw fastq to vcf calls</a></li>
        
            <li><a href="#the-ngseasy-pipelines">The NGSeasy Pipelines</a></li>
        
            <li><a href="#ngs_full_gatk">ngs_full_gatk</a></li>
        
    
        <li class="main "><a href="#output-suffixes">Output suffixes</a></li>
        
            <li><a href="#alignment-output">Alignment Output</a></li>
        
            <li><a href="#thresholds-for-variant-calling-etc">Thresholds for Variant calling etc</a></li>
        
            <li><a href="#gottchas">Gottchas</a></li>
        
            <li><a href="#cnv-tools-to-think-about">CNV tools to think about</a></li>
        
            <li><a href="#useful-links">Useful Links</a></li>
        
    
    </ul>
</div></div>
                <div class="col-md-9" role="main">

<h1 id="ngseasy-beta">NGSeasy (beta)</h1>
<p><strong> This is the latest dev project </strong><br />
<a href="http://core.brc.iop.kcl.ac.uk">Funded by Biomedical Research Centre</a>: http://core.brc.iop.kcl.ac.uk</p>
<p>Publication: pending</p>
<p>Authors: Stephen J Newhouse, Amos Folarin , Maximilian Kerz<br />
Release Version: <strong>1.0.0b</strong>  </p>
<hr />
<h3 id="a-dockerized-ngs-pipeline-and-tool-box">[A <a href="https://www.docker.com/">Dockerized</a> NGS pipeline and tool-box]</h3>
<hr />
<p><strong>With NGSeasy you can now have full suite of NGS tools up and running on any high end workstation in an afternoon</strong></p>
<p><strong>Getting started: <a href="https://github.com/KHP-Informatics/ngs#the-12-easy-steps-to-ngs-freedom">The 12 Easy Steps to NGS Freedom!</a></strong>  </p>
<p><strong>Note:</strong> NGSeasy is under <strong>heavy development</strong> and the code and docs evolve quickly.  </p>
<ul>
<li>
<p><strong>NGSeasy-v1.0 Full Production release will be available Early 2015</strong>  </p>
</li>
<li>
<p><strong>NGSeasy-v1.0.0b Full Production release contains most of the core fucntionality to go from raw fastq to raw vcf calls</strong></p>
</li>
<li><strong>NGSeasy updates every 12 months</strong></li>
<li><strong>GUI in development</strong></li>
<li><strong>Contact us of Reference Genomes and resource files</strong>  </li>
</ul>
<hr />
<blockquote>
<p>NGSeasy is completely open source and we encourage interested folks to jump in and get involved in the dev with us.</p>
</blockquote>
<hr />
<h1 id="ngseasy-easy-analysis-of-next-generation-sequencing">NGSeasy (Easy Analysis of Next Generation Sequencing)</h1>
<p>We present <strong>NGSeasy (Easy Analysis of Next Generation Sequencing)</strong>, a flexible and easy-to-use NGS pipeline for automated alignment, quality control, variant calling and annotation. The pipeline allows users with minimal computational/bioinformatic skills to set up and run an NGS analysis on their own samples, in less than an afternoon, on any operating system (Windows, iOS or Linux) or infrastructure (workstation, cluster or cloud).</p>
<p>NGS pipelines typically utilize a large and varied range of software components and incur a substantial configuration burden during deployment which limits their portability to different computational environments. NGSeasy simplifies this by providing the pipeline components encapsulated in Dockerâ„¢ containers and bundles in a wide choice of tools for each module. Each module of the pipeline represents one functional grouping of tools (e.g. sequence alignment, variant calling etc.).</p>
<p>Deploying the pipeline is as simple as pulling the container images from the public repository into any host running Docker. NGSeasy can be deployed on any medium to high-end workstation, high performance computer cluster and compute clouds (public/private cloud computing) - enabling instant access to elastic scalability without investment overheads for additional compute hardware and makes open and reproducible research straight forward for the greater scientific community.</p>
<h3 id="advantages">Advantages</h3>
<ul>
<li>Easy to use for non-informaticians.  </li>
<li>All run from a single config file that can be made in Excel.  </li>
<li>User can select from mutiple aligners, variant callers and variant annotators</li>
<li>No scary python, .yaml or .json files...just one simple Excel workbook saved as a textfile.  </li>
<li>Just follow our simple set of instructions and NGS away!  </li>
<li>Choice of aligners and variant callers and anntators  </li>
<li>Allows reproducible research  </li>
<li>Version controlled for auditing  </li>
<li>Customisable  </li>
<li>Easy to add new tools  </li>
<li>If it's broke...we will fix it..</li>
<li>Enforced naming convention and directory structures  </li>
<li>Allows users to run "Bake Offs" between tools with ease  </li>
</ul>
<p>We have adapted the current best practices from the Genome Analysis Toolkit (GATK, http://www.broadinstitute.org/gatk/guide/best-practices)  for processing raw alignments in SAM/BAM format and variant calling. The current workflow, has been optimised for Illumina platforms, but can easily be adapted for other sequencing platforms, with minimal effort.  </p>
<p>As the containers themselves can be run as executables with pre-specified cpu and RAM resources, the orchestration of the pipeline can be placed under the control of conventional load balancers if this mode is required.  </p>
<hr />
<h3 id="author-contact-details">Author Contact Details</h3>
<p>Please contact us for help/guidance on using the beta release. </p>
<ul>
<li>
<p>Amos Folarin <a href="./&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#97;&#109;&#111;&#115;&#102;&#111;&#108;&#97;&#114;&#105;&#110;&#64;&#103;&#109;&#97;&#105;&#108;&#46;&#99;&#111;&#109;">&#97;&#109;&#111;&#115;&#102;&#111;&#108;&#97;&#114;&#105;&#110;&#64;&#103;&#109;&#97;&#105;&#108;&#46;&#99;&#111;&#109;</a>  <a href="https://twitter.com/amosfolarin?lang=en">@amosfolarin</a> <br />
<a href="http://www.linkedin.com/pub/amos-folarin/34/b06/978">
<img src="http://www.linkedin.com/img/webpromo/btn_viewmy_160x33.png" width="160" height="33" alt="View Amos's profile on LinkedIn">
</a></p>
</li>
<li>
<p>Stephen J Newhouse <a href="./&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#115;&#116;&#101;&#112;&#104;&#101;&#110;&#46;&#106;&#46;&#110;&#101;&#119;&#104;&#111;&#117;&#115;&#101;&#64;&#103;&#109;&#97;&#105;&#108;&#46;&#99;&#111;&#109;">&#115;&#116;&#101;&#112;&#104;&#101;&#110;&#46;&#106;&#46;&#110;&#101;&#119;&#104;&#111;&#117;&#115;&#101;&#64;&#103;&#109;&#97;&#105;&#108;&#46;&#99;&#111;&#109;</a> <a href="https://twitter.com/s_j_newhouse?lang=en">@s_j_newhouse</a><br />
<a href="http://uk.linkedin.com/pub/dr-stephen-newhouse/29/89a/11a">
<img src="http://www.linkedin.com/img/webpromo/btn_viewmy_160x33.png" width="160" height="33" alt="View Steve's profile on LinkedIn">
</a></p>
</li>
</ul>
<p><strong>Lets us know if you want other tools added to NGSeasy</strong></p>
<p><em>Institution: NIHR Maudsley Biomedical Research Centre For Mental Health and Dementia Unit (Denmark Hill), at The Institute of Psychiatry, Psychology &amp; Neuroscience (IoPPN), Kings College London</em> </p>
<hr />
<h1 id="overview-of-the-ngseasy-pipeline-components">Overview of the NGSeasy Pipeline Components</h1>
<p>The basic pipeline contains all the basic tools needed for manipulation and 
quality control of raw fastq files (ILLUMINA focused), SAM/BAM manipulation,
alignment, cleaning (based on GATK best practises [http://www.broadinstitute.org/gatk/guide/best-practices]) and first pass
variant discovery. Separate containers are provided for indepth variant annotation,
structural variant calling, basic reporting and visualisations.  </p>
<p><img alt="ngsEASY" src="https://github.com/KHP-Informatics/ngs/blob/dev2/figs/ngsEASY_component_visualisation.png" title="Dockerized NGS Pipeline" /></p>
<hr />
<h1 id="the-full-ngseasy-pipeline">The Full NGSeasy pipeline</h1>
<p>The NGSeasy pipelines implement the following :-   </p>
<ul>
<li>
<p><strong>Quality control of raw fastq</strong> files using <strong><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FASTQC</a></strong>  </p>
</li>
<li>
<p><strong>Read trimming</strong> using <strong><a href="http://www.usadellab.org/cms/?page=trimmomatic">TRIMMOMATIC</a></strong>.   </p>
</li>
<li>
<p><strong>Alignment</strong> using one of </p>
<ul>
<li><strong><a href="http://bio-bwa.sourceforge.net/">BWA</a></strong>  </li>
<li><strong><a href="http://www.well.ox.ac.uk/project-stampy">STAMPY</a></strong>   </li>
<li><strong><a href="http://www.novocraft.com">NOVOALIGN</a></strong>  </li>
<li><strong><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">BOWTIE2</a></strong>  </li>
<li><em><a href="http://snap.cs.berkeley.edu/">SNAP</a>: COMING SOON!</em></li>
</ul>
</li>
<li>
<p><strong>SAM/BAM sorting and indexing</strong> with <strong><a href="https://github.com/samtools/samtools">SAMTOOLS</a></strong>.  </p>
</li>
<li>
<p><strong>Read Group information added</strong> using <strong><a href="http://broadinstitute.github.io/picard/">PICARDTOOLS</a>:<a href="http://broadinstitute.github.io/picard/command-line-overview.html#AddOrReplaceReadGroups">AddOrReplaceReadGroups</a></strong> </p>
</li>
<li>
<p><strong>Duplicate marking</strong> with <strong><a href="http://broadinstitute.github.io/picard/">PICARDTOOLS</a>:<a href="http://broadinstitute.github.io/picard/command-line-overview.html#MarkDuplicates">MarkDuplicates</a></strong>.  </p>
</li>
</ul>
<blockquote>
<p>For academic users and/or commercial/clinical groups whom have paid for GATK licensing, the next steps are to perform   </p>
</blockquote>
<ul>
<li><strong>Indel indel realignment and base quality score recalibration</strong> using <strong><a href="https://www.broadinstitute.org/gatk/">GATK</a></strong> built in tools :<ul>
<li><strong><a href="https://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_gatk_tools_walkers_indels_RealignerTargetCreator.php">GATK:RealignerTargetCreator</a></strong> </li>
<li><strong><a href="https://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_gatk_tools_walkers_indels_IndelRealigner.php">GATK:IndelRealigner</a></strong> </li>
<li><strong><a href="https://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_gatk_tools_walkers_bqsr_BaseRecalibrator.php">GATK:BaseRecalibrator</a></strong> </li>
</ul>
</li>
</ul>
<blockquote>
<p>For the non-GATK version    </p>
</blockquote>
<ul>
<li>
<p><strong>Base quality score recalibration</strong> using <a href="http://genome.sph.umich.edu/wiki/BamUtil">BamUtil</a>  </p>
<ul>
<li><strong><a href="http://genome.sph.umich.edu/wiki/BamUtil:_recab">BamUtil:recab</a></strong> </li>
</ul>
</li>
<li>
<p><strong>Post alignment quality control and reporting</strong> is performed usng a number of tools and custom scripts: </p>
<ul>
<li><strong><a href="https://github.com/samtools/samtools">SAMTOOLS:flagstats</a></strong></li>
<li><strong><a href="https://github.com/arq5x/bedtools2">BEDTOOLS:genomecov</a></strong></li>
<li><strong><a href="https://github.com/arq5x/bedtools2">BEDTOOLS:bamtobed</a></strong></li>
<li><strong><a href="http://broadinstitute.github.io/picard/command-line-overview.html#CollectMultipleMetrics">PICARDTOOLS:CollectMultipleMetrics</a></strong>    </li>
<li><strong><a href="http://broadinstitute.github.io/picard/command-line-overview.html#CollectAlignmentSummaryMetrics">PICARDTOOLS:CollectAlignmentSummaryMetrics</a></strong>    </li>
<li><strong><a href="http://broadinstitute.github.io/picard/command-line-overview.html#CollectWgsMetrics">PICARDTOOLS:CollectWgsMetrics</a></strong>    </li>
<li><strong><a href="http://broadinstitute.github.io/picard/command-line-overview.html#CollectTargetedPcrMetrics">PICARDTOOLS:CollectTargetedPcrMetrics</a></strong> (coming soon)    </li>
</ul>
</li>
<li>
<p><strong>SNP and small INDEL</strong> calling using one of the following or a combibation of these tools, if the <code>ensemble</code> method is called using <strong><a href="https://github.com/chapmanb/bcbio.variation">bcbio.variation variant-ensemble</a></strong></p>
<ul>
<li><strong><a href="https://github.com/ekg/freebayes">FREEBAYES</a></strong> </li>
<li><strong><a href="http://www.well.ox.ac.uk/platypus">PLATYPUS</a></strong> </li>
<li><strong><a href="https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_genotyper_UnifiedGenotyper.php">GATK:UnifiedGenotyper</a></strong> </li>
<li><strong><a href="https://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_gatk_tools_walkers_haplotypecaller_HaplotypeCaller.php">GATK:HaplotypeCaller</a></strong> </li>
</ul>
</li>
<li>
<p><strong>Structural Variant (CNV)</strong> calling using one of the following or or a combibation of if the <code>ensemble</code> methods are called:- </p>
<ul>
<li><strong><a href="https://github.com/tobiasrausch/delly">DELLY</a> : in Dev</strong> </li>
<li><strong><a href="https://github.com/arq5x/lumpy-sv/">LUMPY</a>: in Dev</strong></li>
<li><strong><a href="http://www.bioinf.jku.at/software/cnmops/">cn.MOPS</a>: in Dev</strong></li>
<li><strong><a href="https://www.stt.msu.edu/users/hengwang/mHMM.html">m-HMM</a>: in Dev</strong></li>
<li><strong><a href="http://cran.r-project.org/web/packages/ExomeDepth/index.html">ExomeDepth</a>: in Dev</strong></li>
<li><em><a href="http://www-genepi.med.utah.edu/suppl/SLOPE/index.html">SLOPE</a> : COMING SOON!</em></li>
</ul>
</li>
<li>
<p><strong>Variant annotation</strong> using using one of the following or a combibation of if the <code>ensemble</code> methods are called. </p>
<ul>
<li><strong><a href="http://snpeff.sourceforge.net/">SnpEff</a>: in Dev</strong> </li>
<li><strong><a href="http://www.openbioinformatics.org/annovar/">ANNOVAR</a>: in Dev</strong> </li>
<li><strong><a href="http://www.ensembl.org/info/docs/tools/vep/index.html">VEP</a>: in Dev</strong></li>
</ul>
</li>
<li>
<p><strong>Variant reporting</strong> using custom scripts</p>
</li>
</ul>
<p><strong>Note</strong> Some of the later functions i.e. variant annotation and qc reporting are still in dev.  </p>
<hr />
<p><strong>We highly recommed read trimming prior to alignment.</strong>
We have noticed considerable speed-ups in alignmnet time and increased quality of SNP/INDEL calls using trimmed vs raw fastq.  </p>
<p><strong>Base quality score recalibration is also recommended.</strong><br />
As an alternative to GATK, we have added fucntionality for use of 
<strong><a href="https://github.com/statgen/bamUtil">BamUtil</a>:<a href="http://genome.sph.umich.edu/wiki/BamUtil:_recab">recab</a></strong> 
for base quality score recalibration.</p>
<p><strong>Non-GATK users</strong> 
- are encouraged to use aligners such as <strong><a href="http://www.well.ox.ac.uk/project-stampy">stampy</a></strong> and <strong><a href="http://www.novocraft.com">novoalign</a></strong> that perform base quality score recal on the fly.<br />
- are encouraged to use variant callers that perform local re-aligmnet around candidate sites to mitigate the need for the indel realignment stages.<br />
    - <strong><a href="https://github.com/ekg/freebayes">freebayes</a></strong>
    - <strong><a href="http://www.well.ox.ac.uk/platypus">platypus</a></strong></p>
<hr />
<h3 id="coming-soon">Coming Soon</h3>
<ul>
<li>New Aligners:- <a href="http://snap.cs.berkeley.edu/">SNAP</a>, GSNAP, mr- and mrs-Fast,gem</li>
<li>https://github.com/amplab/snap</li>
<li>[SLOPE (CNV fo targetted NSG)] ((http://www.biomedcentral.com/1471-2164/12/184)) </li>
<li>Cancer Pipelines</li>
<li>Annotation Pipelines and Databases</li>
<li>Visualisation Pipelines</li>
<li>Var Callers:- VarScan2</li>
<li>SGE scripts and basic BASH scrips for running outside of Docker</li>
<li>biobambam https://github.com/gt1/biobambam  </li>
<li>bamaddrg https://github.com/ekg/bamaddrg  </li>
<li>bamtools https://github.com/ekg/bamtools  </li>
</ul>
<hr />
<h1 id="dockerised-ngseasy">Dockerised NGSeasy</h1>
<p><img alt="docker" src="https://github.com/KHP-Informatics/ngs/blob/master/figs/Docker_container_engine_logo.png" title="Docker" />  </p>
<p>The following section describes getting the Dockerised NGSeasy Pipeline(s) and Resources, project set up and running NGSeasy in <strong>12 easy steps</strong>.  </p>
<p>Getting all resources and building required tools will take a few hours depending on network connections and any random "ghosts in the machine".</p>
<hr />
<h2 id="the-12-easy-steps-to-ngs-freedom">The 12 Easy Steps to NGS Freedom!</h2>
<p><strong>1. <a href="https://github.com/KHP-Informatics/ngs#1-install-docker">Install Docker</a></strong><br />
<strong>2. <a href="https://github.com/KHP-Informatics/ngs#2-get-ngseasy-automated-build-container-images">Get NGSeasy Automated build Container Images</a></strong><br />
<strong>3. <a href="https://github.com/KHP-Informatics/ngs#3-make-ngseasy-project-directory">Make NGSeasy project directory</a></strong><br />
<strong>4. <a href="https://github.com/KHP-Informatics/ngs#4-download-ngseasy-resources">Download NGSeasy Resources</a></strong><br />
<strong>5. <a href="https://github.com/KHP-Informatics/ngs#5-get-ngseasy-sripts">Get NGSeasy Sripts</a></strong><br />
<strong>6. <a href="https://github.com/KHP-Informatics/ngs#6-manually-build-required-ngseasy-container-images">Manually Build required NGSeasy Container Images</a></strong><br />
<strong>7. <a href="https://github.com/KHP-Informatics/ngs#7-manually-build-ngseasy-variant-annotaion-container-images">Manually Build NGSeasy Variant Annotaion Container Images</a></strong><br />
<strong>8. <a href="https://github.com/KHP-Informatics/ngs#8-set-up-ngseasy-project-working-directories">Set up NGSeasy Project Working Directories</a></strong><br />
<strong>9. <a href="https://github.com/KHP-Informatics/ngs#9-ngseasy-project-configuration-file">NGSeasy Project configuration file</a></strong><br />
<strong>10. <a href="https://github.com/KHP-Informatics/ngs#10-copy-project-fastq-files-to-relevent-projectsample-directories">Copy Project Fastq files to relevent Project/Sample Directories</a></strong><br />
<strong>11. <a href="https://github.com/KHP-Informatics/ngs#11-start-the-ngseasy-volume-contaier">Start the NGSeasy Volume Contaier</a></strong><br />
<strong>12. <a href="https://github.com/KHP-Informatics/ngs#12-running-an-ngseasy-full-pipeline--from-raw-fastq-to-vcf-calls">Running an NGSeasy full pipeline : from raw fastq to vcf calls</a></strong>  </p>
<hr />
<h2 id="1-install-docker">1. Install Docker</h2>
<p>Follow the simple instructions in the links provided below  </p>
<ul>
<li><a href="https://docs.docker.com/installation/mac/">Mac</a>  </li>
<li><a href="https://docs.docker.com/installation/windows/">Windows</a></li>
<li><a href="https://docs.docker.com/installation/ubuntulinux/">Ubuntu</a></li>
</ul>
<p>A full set of instructions for multiple operating systems are available on the <a href="https://docs.docker.com/installation/">Docker website</a>.</p>
<h2 id="2-get-ngseasy-automated-build-container-images">2. Get NGSeasy Automated build Container Images</h2>
<p>All NGSeasy Docker images can be pulled down from <strong><a href="https://hub.docker.com/u/compbio/">compbio Docker Hub</a></strong> or using the script <a href="https://github.com/KHP-Informatics/ngs/blob/master/bin/get_containers.sh">get_containers.sh</a></p>
<pre><code class="bash"># get images
bash get_containers.sh v1.0
</code></pre>

<h3 id="dockerised-and-automated-builds">Dockerised and Automated Builds</h3>
<p>The following opensource tools are all provided as automated builds. </p>
<table>
<thead>
<tr>
<th>Tool</th>
<th>Build</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-base/">ngseasy-base</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-fastqc">fastqc</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-trimmomatic">trimmomatic</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-bwa">bwa</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-bowtie">bowtie</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-picardtools">picardtools</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-samtools">samtools</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-freebayes/">freebayes</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-bedtools/">bedtools</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-bcbiovar/">bcbiovar</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-delly">delly</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-lumpy">lumpy</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-cnmops">cnmops</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-mhmm">mhmm</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-exomedepth">exomedepth</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-bamutil">bamutil</a></td>
<td>automated build</td>
</tr>
</tbody>
</table>
<p>samtools includes bcftools and htslib  </p>
<p>Its as easy as: - </p>
<pre><code class="bash">docker pull compbio/ngseasy-${TOOL}
</code></pre>

<hr />
<h2 id="3-make-ngseasy-project-directory">3. Make NGSeasy project directory</h2>
<p>The user needs to make the relevent directory structures on their local machine before starting an NGS run. </p>
<p>On our sysetm we typically set up a top-level driectory called <strong>ngs_projects</strong> within which we store output from all our individual NGS projects. </p>
<p>Here we are working from local top level directory called <strong>media/</strong>, but this can really be any folder on your local system ie your home directory <strong>~/${USER}</strong>.  </p>
<p>Within this directory <strong>media</strong> we make the following folders: - </p>
<pre><code class="bash">ngs_projects  
|  
|__raw_fastq  
|__config_files  
|__reference_genomes_b37  
|__gatk_resources  
|__ngseasy
</code></pre>

<p><strong>Note</strong> The following directories are obtained in step <strong><a href="https://github.com/KHP-Informatics/ngs/tree/master#4-download-ngseasy-resources">4. Download NGSeasy Resources</a></strong>.<br />
<strong>- reference_genomes_b37  </strong><br />
<strong>- gatk_resources </strong>  </p>
<p><strong>Move to media</strong></p>
<pre><code class="bash"># Move to media/
cd media
</code></pre>

<p><strong>make toplevel ngs_projects folder</strong></p>
<pre><code class="bash"># make toplevel NGS folder
mkdir ngs_projects 
</code></pre>

<p><strong>make fast_raw folder</strong></p>
<pre><code class="bash"># fastq staging area
mkdir ngs_projects/fastq_raw 
</code></pre>

<p><strong>make config_files folder</strong></p>
<pre><code class="bash"># config files
mkdir ngs_projects/config_files 
</code></pre>

<p><strong>make ngseasy folder</strong></p>
<pre><code class="bash"># NGSeasy scripts
mkdir ngs_projects/ngseasy 
</code></pre>

<hr />
<h2 id="4-download-ngseasy-resources">4. Download NGSeasy Resources</h2>
<p>Download the indexed reference genomes and example data for use with NGSeasy.</p>
<p><strong>NGSeasy Resources:-</strong><br />
- <strong>reference_genomes_b37.tgz</strong> b37 reference genomes indexed for use with all provided aligners (BWA, Bowtie2, Stampy, Novoalign) and annotation bed files for use with pipeline scripts
- <strong>gatk_resources.tar.gz</strong> gatk resources bundle
- <strong>fastq_example.tgz</strong> Example 75bp PE Illumina Whole Exome Sequence fastq data for <strong>NA12878</strong>
- Annotation Databases Coming in the next update </p>
<p><strong>Download the data to the top level directory</strong></p>
<h3 id="ftp-details">FTP Details</h3>
<ul>
<li><strong>ftp:</strong>  159.92.120.21  </li>
<li><strong>user:</strong> compbio-public  </li>
<li><strong>pwd:</strong>  compbio-public  </li>
<li><strong>port:</strong> 21  </li>
</ul>
<p><strong>Move to top level directory</strong></p>
<pre><code class="bash">cd ngs_projects
</code></pre>

<p><strong>FTP NGSeasy Resources</strong></p>
<pre><code class="bash">ftp 159.92.120.21
</code></pre>

<p><strong>mget NGSeasy Resources</strong></p>
<pre><code class="bash">ftp&gt; cd /Public/NGSeasy_Public_Resources
ftp&gt; prompt off
ftp&gt; mget *.gz
ftp&gt; exit
</code></pre>

<p>I would recommend using a separate program like <a href="https://filezilla-project.org/">FileZilla</a>, which will make it much easier for you to set up and manage your file transfers</p>
<p><strong>Extract NGSeasy Resources</strong></p>
<pre><code class="bash"># Extract resources
cd ngs_projects/
tar xvf gatk_resources.tgz; 
cd ngs_projects/gatk_resources
gunzip *
</code></pre>

<p><strong>Extract NGSeasy Reference Genomes</strong></p>
<pre><code class="bash"># Extract Reference Genomes
cd ngs_projects/
tar xvf reference_genomes_b37.tgz; 
cd ngs_projects/reference_genomes_b37
gunzip *
</code></pre>

<hr />
<h3 id="gatk-resources">GATK Resources</h3>
<ul>
<li>https://www.broadinstitute.org/gatk/guide/article.php?id=1215  </li>
<li>https://www.broadinstitute.org/gatk/guide/article.php?id=1213  </li>
</ul>
<p><strong>Downloading</strong></p>
<pre><code>location: ftp.broadinstitute.org
username: gsapubftp-anonymous
password: &lt;blank&gt;
</code></pre>

<p><strong>b37 Resources: the Standard Data Set</strong>
- Reference sequence (standard 1000 Genomes fasta) along with fai and dict files
- dbSNP in VCF. This includes two files:
    - The most recent dbSNP release
    This file subsetted to only sites discovered in or before dbSNPBuildID 129, which excludes the impact of the 1000 Genomes project and is useful for evaluation of dbSNP rate and Ti/Tv values at novel sites.
    - HapMap genotypes and sites VCFs
- OMNI 2.5 genotypes for 1000 Genomes samples, as well as sites, VCF
    The current best set of known indels to be used for local realignment (note that we don't use dbSNP for this anymore); use both files:
- 1000G_phase1.indels.b37.vcf (currently from the 1000 Genomes Phase I indel calls)
- Mills_and_1000G_gold_standard.indels.b37.sites.vcf
- A large-scale standard single sample BAM file for testing:
    - NA12878.HiSeq.WGS.bwa.cleaned.recal.hg19.20.bam containing ~64x reads of NA12878 on chromosome 20
The results of the latest UnifiedGenotyper with default arguments run on this data set (NA12878.HiSeq.WGS.bwa.cleaned.recal.hg19.20.vcf)</p>
<hr />
<h2 id="5-get-ngseasy-sripts">5. Get NGSeasy Sripts</h2>
<p>We then need to get the latest NGSeasy scripts from <a href="https://github.com/KHP-Informatics/ngs">GitHub</a> . The user is required to download the scripts to the <code>ngseasy</code> directory</p>
<p><strong>move to the <code>ngseasy</code> directory</strong></p>
<pre><code class="bash">cd /media/ngs_projects/nsgeasy
</code></pre>

<p><strong>clone the <a href="https://github.com/KHP-Informatics/ngs">ngs</a> repository</strong></p>
<pre><code class="bash">git clone https://github.com/KHP-Informatics/ngs.git
</code></pre>

<p><strong>add <code>nsgeasy/ngs/bin</code> to your system PATH</strong></p>
<pre><code class="bash">export PATH=$PATH:/media/ngs_projects/nsgeasy/ngs/bin
</code></pre>

<p><strong>or add to global .bashrc</strong></p>
<pre><code class="bash">echo &quot;export PATH=$PATH:/media/ngs_projects/nsgeasy/ngs/bin&quot; ~/.bashrc
source ~/.bashrc
</code></pre>

<p><strong>alternatively donwload the scripts from our <a href="https://github.com/KHP-Informatics/ngs">GitHub Release</a></strong> </p>
<hr />
<h2 id="6-manually-build-required-ngseasy-container-images">6. Manually Build required NGSeasy Container Images</h2>
<p>Currently we are not able to automatically build some of the tools in pre-built docker containers due to licensing restrictions. </p>
<p>Some of the software has restrictions on use particularly for commercial 
purposes. Therefore if you wish to use this for commercial purposes, then you 
leagally have to approach the owners of the various components yourself!  </p>
<p><strong>Software composing the pipeline requiring registration:-</strong>  </p>
<ul>
<li>novoalign http://www.novocraft.com/  </li>
<li>Stampy http://www.well.ox.ac.uk/project-stampy  </li>
<li>Platypus http://www.well.ox.ac.uk/platypus  </li>
<li>GATK https://www.broadinstitute.org/gatk/  </li>
<li>ANNOVAR http://www.openbioinformatics.org/annovar/  </li>
</ul>
<p><strong>These tools require manual download and registration with the proivder. For non-academics/commercial groups, you will need to pay for some of these tools.</strong></p>
<h3 id="dockerised-and-manual-builds">Dockerised and Manual Builds</h3>
<table>
<thead>
<tr>
<th>Tool</th>
<th>Build</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_novoalign">novoalign</a></td>
<td>manual build</td>
</tr>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_annovar">annovar</a></td>
<td>manual build</td>
</tr>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_stampy">stampy</a></td>
<td>manual build</td>
</tr>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/nsgeasy_platypus">platypus</a></td>
<td>manual build</td>
</tr>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_gatk">gatk</a></td>
<td>manual build</td>
</tr>
</tbody>
</table>
<p>Once you have paid/registered and downloaded the tool, we provide scripts and guidance for building these tools on your system.  </p>
<p>Its as easy as:-  </p>
<pre><code class="bash">docker build -t compbio/ngseasy-${TOOL} .
</code></pre>

<hr />
<h3 id="61-building-stampy">6.1 Building Stampy</h3>
<p><strong>resister at http://www.well.ox.ac.uk/project-stampy</strong>  </p>
<p>Download stampy to local directory and check version number. If this differs from the <a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_stampy/Dockerfile">Dockerfile</a> build file, 
then edit the <a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_stampy/Dockerfile">Dockerfile</a> if needed. 
You will be emailed a URL to download stampy. Insert this into the <a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_stampy/Dockerfile">Dockerfile</a></p>
<pre><code class="bash"># on our local system we cd to media
cd /media

# them move to ngs_projects toplevel directory
cd ngs_projects

# and then the ngseasy folder with all our ngs scripts
# git  clone https://github.com/KHP-Informatics/ngs.git
# if you havent alreay
cd ngseasy

# move to ngseasy_stampy folder
cd ngs/ngs_docker_debian/ngseasy_stampy

# build
docker build -t compbio/ngseasy-stampy:v1.0 .
</code></pre>

<hr />
<h3 id="62-building-platypus">6.2 Building Platypus</h3>
<p><strong>resister at http://www.well.ox.ac.uk/platypus</strong></p>
<p>Download platypus to local directory and check version number. If this differs from the <a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/nsgeasy_platypus/Dockerfile">Dockerfile</a> build file, 
then edit the <a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/nsgeasy_platypus/Dockerfile">Dockerfile</a> if needed. 
You will be emailed a URL to download platypus. Insert this into the <a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/nsgeasy_platypus/Dockerfile">Dockerfile</a> </p>
<pre><code class="bash"># on our local system we cd to media
cd /media

# them move to ngs_projects toplevel directory
cd ngs_projects

# and then the ngseasy folder with all our ngs scripts
# git  clone https://github.com/KHP-Informatics/ngs.git
# if you havent alreay
cd ngseasy

# move to ngseasy_stampy folder
cd ngs/ngs_docker_debian/ngseasy_platypus

# build
docker build -t compbio/ngseasy-platypus:v1.0 .
</code></pre>

<hr />
<h3 id="63-building-novoalign">6.3 Building NOVOALIGN</h3>
<p><strong>Download Novoalign from  http://www.novocraft.com/</strong> into the local build directory <strong>ngs/ngs_docker_debian/ngseasy_novoalign</strong>. 
Edit the <a href="https://github.com/KHP-Informatics/ngs/blob/master/containerized/ngs_docker_debian/ngseasy_novoalign/Dockerfile">Dockerfile</a> to relfect
the correct version of novoalign.  </p>
<p>To use all novoalign fucntionality, you will need to <strong>pay for a license</strong>.   </p>
<p>Once you obtained your <strong>novoalign.lic</strong>, download this to the build directory <strong>ngs/ngs_docker_debian/ngseasy_novoalign</strong>, which now should contain your updated <a href="https://github.com/KHP-Informatics/ngs/blob/master/containerized/ngs_docker_debian/ngseasy_novoalign/Dockerfile">Dockerfile</a>.</p>
<pre><code class="bash"># on our local system we cd to media
cd /media

# them move to ngs_projects toplevel directory
cd ngs_projects

# and then the ngseasy folder with all our ngs scripts
# git  clone https://github.com/KHP-Informatics/ngs.git
# if you havent alreay
cd ngseasy

# move to ngseasy_stampy folder
cd ngs/ngs_docker_debian/ngseasy_novoalign
ls 
</code></pre>

<p><strong>the directory should contain the following:-</strong></p>
<pre><code>Dockerfile
novoalign.lic
README.md
novosortV1.03.01.Linux3.0.tar.gz
novocraftV3.02.08.Linux3.0.tar.gz
</code></pre>

<p><strong>build novoalign</strong></p>
<pre><code class="bash"># build
docker build -t compbio/ngseasy-novoalign:v1.0 .
</code></pre>

<hr />
<h3 id="64-building-gatk">6.4 Building GATK</h3>
<p>You need to register and accept the GATK license agreement at https://www.broadinstitute.org/gatk/.  </p>
<p>Once done, download GATK and place in the GTAK build directory <strong>ngs/ngs_docker_debian/ngseasy_gatk</strong>.  </p>
<p>Edit the <a href="https://github.com/KHP-Informatics/ngs/blob/master/containerized/ngs_docker_debian/ngseasy_gatk/Dockerfile">Dockerfile</a> to relfect
the correct version of GATK.  </p>
<pre><code class="bash"># on our local system we cd to media
cd /media

# them move to ngs_projects toplevel directory
cd ngs_projects

# and then the ngseasy folder with all our ngs scripts
# git  clone https://github.com/KHP-Informatics/ngs.git
# if you havent alreay
cd ngseasy

# move to ngseasy_stampy folder
cd ngs/ngs_docker_debian/ngseasy_gatk
ls 
</code></pre>

<p><strong>the directory should contain the following:-</strong></p>
<pre><code>Dockerfile
README.md
GenomeAnalysisTK-3.3-0.tar.bz2
</code></pre>

<p><strong>build gatk</strong></p>
<pre><code class="bash"># build
docker build -t compbio/ngseasy-gatk:v1.0 .
</code></pre>

<hr />
<h2 id="7-manually-build-ngseasy-variant-annotaion-container-images">7. Manually Build NGSeasy Variant Annotaion Container Images</h2>
<p>The tools used for variant annotation use large databases and the docker images exceed 10GB. Therefore, the user should manually build these container images prior to running the NGS pipelines.
Docker build files (<a href="https://docs.docker.com/jsearch/?q=Dockerfile">Dockerfile</a>) are available for 
- <a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_annovar/Dockerfile">Annovar</a><br />
- <a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_vep/Dockerfile">VEP</a> <br />
- <a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_snpeff/Dockerfile">snpEff</a>  </p>
<p><strong>Note</strong> Annovar requires user registration.  </p>
<p><strong>Once built on the user system, these container images can persist for as long as the user wants.</strong>  </p>
<p><strong>Large Variant Annotation Container Images</strong></p>
<table>
<thead>
<tr>
<th>Tool</th>
<th>Build</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_annovar">annovar</a></td>
<td>manual build</td>
</tr>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_vep">vep</a></td>
<td>manual build</td>
</tr>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_snpeff">snpeff</a></td>
<td>manual build</td>
</tr>
</tbody>
</table>
<p>Its as easy as:-  </p>
<pre><code class="bash">docker build -t compbio/ngseasy-${TOOL} .
</code></pre>

<hr />
<h3 id="71-build-vep">7.1 Build VEP</h3>
<pre><code class="bash">
cd /media/ngs_projects/nsgeasy/ngs/containerized/ngs_docker_debian/ngseasy_vep

sudo docker build -t compbio/ngseasy-vep:${VERSION} .
</code></pre>

<hr />
<h3 id="72-build-annovar">7.2 Build Annovar</h3>
<pre><code class="bash">cd /media/ngs_projects/nsgeasy/ngs/containerized/ngs_docker_debian/ngseasy_annovar

sudo docker build -t compbio/ngseasy-annovar:${VERSION} .
</code></pre>

<hr />
<h3 id="73-build-snpeff">7.3 Build snpEff</h3>
<pre><code class="bash">cd /media/ngs_projects/nsgeasy/ngs/containerized/ngs_docker_debian/ngseasy_snpeff

sudo docker build -t compbio/ngseasy-snpeff:${VERSION} .
</code></pre>

<hr />
<h2 id="8-set-up-ngseasy-project-working-directories">8. Set up NGSeasy Project Working Directories</h2>
<p>Running the script <code>ngseasy_initiate_project</code> ensures that all relevant directories are set up, and also enforces a clean structure to the NGS project.  </p>
<p>Within this we make a <code>raw_fastq</code> folder, where we temporarily store all the raw fastq files for each project. 
This folder acts as an initial stagging area for the raw fastq files. During the project set up, we copy/move project/sample related fastq files to their own specific directories.
Fastq files must have suffix and be gzipped: <strong>_1.fq.gz</strong> or <strong>_2.fq.gz</strong><br />
furture version will allow any format  </p>
<p>Running <code>ngseasy_initiate_project</code> with the relevent configuration file, will set up the following directory structure for every project and sample within a project:-  </p>
<h2 id="ngs-project-directory">NGS Project Directory</h2>
<pre><code class="bash">.
ngs_projects  
|  
|__raw_fastq  
|__config_files  
|__reference_genomes_b37  
|__gatk_resources  
|__ngseasy
|
|__ project_id  
    |  
    |__run_logs  
    |__config_files  
    |__project_vcfs  
    |__project_bams  
    |__project_reports  
    |
    |__sample_id_1  
    |   |  
    |   |__fastq  
    |   |__tmp  
    |   |__alignments  
    |   |__vcf  
    |   |__reports  
    |   |__config_files  
    |
    |
    |__sample_id_n  
        |  
        |__fastq  
        |__tmp  
        |__alignments  
        |__vcf  
        |__reports  
        |__config_files  
</code></pre>

<p><strong>Running <code>ngseasy_initiate_project</code></strong></p>
<pre><code class="bash">ngseasy_initiate_project -c config.file.tsv -d /media/ngs_projects
</code></pre>

<hr />
<h2 id="9-ngseasy-project-configuration-file">9. NGSeasy Project configuration file</h2>
<p>In Excel make config file and save as [TAB] Delimited file with <code>.tsv</code> extenstion.<br />
See Example provided and <a href="https://docs.google.com/spreadsheets/d/1kp1Nyw0x3zXqO2Wm2Z25ErJ0Z-Uoab8tjRPq9h4sonk/edit?usp=sharing">GoogleDoc</a>. Remove the header from this file before running the pipeline. This sets up Information related to: Project Name, Sample Name, Library Type, Pipeline to call, NCPU.</p>
<p>The [config.file.tsv] should contain the following 15 columns for each sample to be run through a pipeline:- </p>
<table>
<thead>
<tr>
<th>Variable</th>
<th>type</th>
<th>Description</th>
<th>Options/Examples</th>
</tr>
</thead>
<tbody>
<tr>
<td>POJECT_ID</td>
<td>string</td>
<td>Project ID</td>
<td>Cancer</td>
</tr>
<tr>
<td>SAMPLE_ID</td>
<td>string</td>
<td>Sample ID</td>
<td>T100</td>
</tr>
<tr>
<td>FASTQ1</td>
<td>string</td>
<td>Raw fastq file name read 1</td>
<td>foo_1_fq.gz</td>
</tr>
<tr>
<td>FASTQ2</td>
<td>string</td>
<td>Raw fastq file name read 1</td>
<td>foo_2_fq.gz</td>
</tr>
<tr>
<td>PROJECT_DIR</td>
<td>string</td>
<td>Project Directory</td>
<td>/medida/ngs_projects</td>
</tr>
<tr>
<td>DNA_PREP_LIBRARY_ID</td>
<td>string</td>
<td>DNA Libray Prep ID</td>
<td>Custom_Cancer</td>
</tr>
<tr>
<td>NGS_PLATFORM</td>
<td>string</td>
<td>Platform Name</td>
<td>ILLUMINA</td>
</tr>
<tr>
<td>NGS_TYPE</td>
<td>string</td>
<td>Experiment type</td>
<td>WGS/WEX/TGS/</td>
</tr>
<tr>
<td>BED_ANNO</td>
<td>string</td>
<td>Annotation Bed File</td>
<td>exons_b37.bed</td>
</tr>
<tr>
<td>PIPELINE</td>
<td>string</td>
<td>NGSeasy Pipeline Script</td>
<td>ngs_full_gatk/ngs_full_no_gatk</td>
</tr>
<tr>
<td>ALIGNER</td>
<td>string</td>
<td>Aligner</td>
<td>bwa/bowtie/stampy/novoalign</td>
</tr>
<tr>
<td>VARCALLER</td>
<td>string</td>
<td>Variant Caller</td>
<td>ensemble/freebayes/platypus/UnifiedGenotyper/HaplotypeCaller</td>
</tr>
<tr>
<td>GTMODEGATK</td>
<td>string</td>
<td>GATK Variant Caller Mode</td>
<td>EMIT_ALL_CONFIDENT_SITES/EMIT_VARIANTS_ONLY</td>
</tr>
<tr>
<td>CLEANUP</td>
<td>string</td>
<td>Clean Up Files (TRUE/FALSE)</td>
<td>TRUE/FALSE</td>
</tr>
<tr>
<td>NCPU</td>
<td>number</td>
<td>Number of cores to call</td>
<td>1..n</td>
</tr>
<tr>
<td>VERSION</td>
<td>number</td>
<td>NGSeasy Version</td>
<td>v0.9/v1.0</td>
</tr>
</tbody>
</table>
<p>In the config file we set PIPELINE to call the pipeline <strong>[ngs_full_gatk]</strong> or <strong>[ngs_full_no_gatk]</strong>.   </p>
<p><em>coming soon</em> options to add user email, specify non-gatk runs  </p>
<hr />
<h2 id="10-copy-project-fastq-files-to-relevent-projectsample-directories">10. Copy Project Fastq files to relevent Project/Sample Directories</h2>
<pre><code class="bash">ngseasy_initiate_fastq -c config.file.tsv -d /media/ngs_projects
</code></pre>

<hr />
<h2 id="11-start-the-ngseasy-volume-contaier">11. Start the NGSeasy Volume Contaier</h2>
<p>In the Docker container the project directory is mounted in <code>/home/pipeman/ngs_projects</code></p>
<pre><code class="bash">ngseasy_volumes_container -d /media/ngs_projects
</code></pre>

<p><strong>inside ngseasy_volumes_container</strong>. This is what it is calling. Note the directory names and mounts. </p>
<pre><code class="bash">
# host_vol_dir = ngs_projects

  docker run \
  -d \
  -P \
  -v ${host_vol_dir}/fastq_raw:/home/pipeman/fastq_raw \
  -v ${host_vol_dir}/reference_genomes_b37:/home/pipeman/reference_genomes_b37 \
  -v ${host_vol_dir}/gatk_resources:/home/pipeman/gatk_resources \
  -v ${host_vol_dir}:/home/pipeman/ngs_projects \
  -v ${host_vol_dir}/ngseasy/ngs/bin:/home/pipeman/ngseasy_scripts \
  --name volumes_container \
  -t compbio/ngseasy-base:wheezy
</code></pre>

<hr />
<h2 id="12-running-an-ngseasy-full-pipeline-from-raw-fastq-to-vcf-calls">12. Running an NGSeasy full pipeline : from raw fastq to vcf calls</h2>
<p><strong>run ngseay</strong></p>
<pre><code class="bash">
    ngseasy -c config.file.tsv -d /media/nsg_projects

</code></pre>

<p>The pipeline is defined in the config file as <strong>[ngs_full_gatk]</strong></p>
<hr />
<h2 id="the-ngseasy-pipelines">The NGSeasy Pipelines</h2>
<table>
<thead>
<tr>
<th>Pipeline</th>
<th>Short Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/blob/master/bin/ngs_full_gatk">ngs_full_gatk</a></td>
<td>fastq to recalibrated bam to vcf using GATK</td>
</tr>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/blob/master/bin/ngs_full_no_gatk">ngs_full_no_gatk</a></td>
<td>fastq to recalibrated bam to vcf</td>
</tr>
</tbody>
</table>
<p>gatk version includes indel realignment and base recalibration.  </p>
<p>Non-academics/commercial groups need to pay for GATK.  </p>
<p>Currently <strong>ngs_full_gatk</strong> pipeline is the most developed module.  </p>
<p>The <strong>ngs_full_no_gatk</strong> pipeline provides alternatives to processing with GATK. Here BamUtil:recab is used to recalibrate base quality scores and freebayes/platypus are the variant callers of choice.</p>
<h2 id="ngs_full_gatk">ngs_full_gatk</h2>
<p>Each <strong>pipeline</strong> is a bash wrapper that calls a number of functions/steps set out in <a href="https://github.com/KHP-Informatics/ngs#the-full-ngseasy-pipeline">The Full NGSeasy pipeline</a>.</p>
<p>Here  <strong>[ngs_full_gatk]</strong> is a wrapper/fucntion for calling an NGS pipeline. The inside to this script is set out below:-  </p>
<pre><code class="bash">#!/bin/bash -x

#usage printing func
usage()
{
cat &lt;&lt; EOF
  This script calls the NGSeasy pipeline ngs_full_gatk

  ARGUMENTS:
  -h      Flag: Show this help message
  -c      NGSeasy project and run configureation file
  -d      NGSeasy project directory

  EXAMPLE USAGE:

    ngseasy -c config.file.tsv -d project_directory

EOF
}

#get options for command line args
  while  getopts &quot;hc:d:&quot; opt
  do

      case ${opt} in
      h)
      usage #print help
      exit 0
      ;;

      c)
      config_tsv=${OPTARG}
      ;;

      d)
      project_directory=${OPTARG}
      ;; 
      esac
  done

#check config file exists.
if [ ! -e &quot;${config_tsv}&quot; ] 
then
        echo &quot;ERROR :  ${config_tsv} does not exist &quot;
        usage;
        exit 1;
fi

#check exists.
  if [ ! -d &quot;${project_directory}&quot; ] 
  then
      echo &quot; ERROR : ${project_directory} does not exist &quot;
      usage;
      exit 1;
  fi

##################  
# start pipeline #
##################

# Each of these fucntions will call the required image/container(s) and run a part of the NGS pipeline. Each step is usually 
# dependent on the previous step(s) - in that they require certain data/input/output in the correct format 
# and with the correct nameing conventions enforced by our pipeline to exist, before executing.

ngseasy_fastqc  -c ${config_tsv} -d ${project_directory}

ngseasy_trimmomatic -c ${config_tsv} -d ${project_directory}

ngseasy_alignment -c ${config_tsv} -d ${project_directory}

ngseasy_addreadgroup -c ${config_tsv} -d ${project_directory}

ngseasy_markduplicates -c ${config_tsv} -d ${project_directory}

ngseasy_indel_realn -c ${config_tsv} -d ${project_directory}

ngseasy_base_recal -c ${config_tsv} -d ${project_directory}

ngseasy_filter_recalbam -c ${config_tsv} -d ${project_directory}

ngseasy_alignment_qc -c ${config_tsv} -d ${project_directory}

ngseasy_variant_calling -c ${config_tsv} -d ${project_directory}

# coming soon...
# ngseasy_filter_bam -c ${config_tsv} -d ${project_directory}
# ngseasy_cnv_calling -c ${config_tsv} -d ${project_directory}
# ngseasy_variant_filtering -c ${config_tsv} -d ${project_directory}
# ngseasy_variant_annotation -c ${config_tsv} -d ${project_directory}
# ngseasy_report -c ${config_tsv} -d ${project_directory}
</code></pre>

<hr />
<h1 id="output-suffixes">Output suffixes</h1>
<h3 id="alignment-output">Alignment Output</h3>
<p><em>.raw.sam  (WEX ~ 8GB)
</em>.raw.bam<br />
<em>.raw.bai<br />
</em>.sort.bam  (WEX ~ 3GB)
*.sort.bai  </p>
<hr />
<h3 id="addreadgroup">Addreadgroup</h3>
<p><em>.addrg.bam<br />
</em>.addrg.bai<br />
*.addrg.bam.bai  </p>
<hr />
<h3 id="dupemark">Dupemark</h3>
<p><em>.dupemk.bam<br />
</em>.dupemk.bai
*.dupemk.bam.bai  </p>
<hr />
<h3 id="indel-realign">Indel realign</h3>
<p><em>.realn.bam<br />
</em>.realn.bai
*.realn.bam.bai  </p>
<hr />
<h3 id="base-recal">Base recal</h3>
<p><em>.recal.bam  (WEX ~ 4.4G)<br />
</em>.recal.bai<br />
<em>.recal.bam.bai<br />
</em>.realn.bam.BaseRecalibrator.table<br />
<em>.recal.bam.BaseRecalibrator.table<br />
</em>.recal.bam.BaseRecalibrator.BQSR.csv  </p>
<hr />
<h2 id="thresholds-for-variant-calling-etc">Thresholds for Variant calling etc</h2>
<p>For Freebayes and Platypus tools:-  </p>
<ul>
<li>We set min coverage to 10  </li>
<li>Min mappinng quality to 20  </li>
<li>Min base quality to 20</li>
</ul>
<p>For GATK HaplotypeCaller (and UnifiedGenotyper)</p>
<p><code>-stand_call_conf 30 -stand_emit_conf 10 -dcov 250 -minPruning 10</code></p>
<p>Note: <code>minPruning 10</code> was added as many runs of HaplotypeCaller failed when using non-bwa aligend and GATK best practices cleaned BAMs. This fix sorted all problems out, and you really dont want dodgy variant calls...do you? Same goes for thresholds hard coded for use with Freebayes and Platypus.<br />
These setting all work well in our hands. Feel  free to edit the scripts to suit your needs.</p>
<hr />
<p>blah blah blah</p>
<hr />
<h2 id="gottchas">Gottchas</h2>
<p><strong>bin/bash -c</strong></p>
<ul>
<li>need to add <code>/bin/bash -c ${COMMAND}</code> when software require <code>&gt;</code> redirect to some output</li>
</ul>
<p>example below for bwa:-  </p>
<pre><code>  sudo docker run \
  -P \
  --name sam2bam_${SAMPLE_ID} \
  --volumes-from volumes_container \
  -t compbio/ngseasy-samtools:v0.9 /bin/bash -c \
  &quot;/usr/local/pipeline/samtools/samtools view -bhS ${SOUTDocker}/alignments/${BAM_PREFIX}.raw.bwa.sam &gt; ${SOUTDocker}/alignments/${BAM_PREFIX}.raw.bwa.bam&quot;
  ```

runnig this without ```/bin/bash -c``` breaks. The ```&gt;``` is called outside of the container

### The Annoying thing about GATK!
This will break your runs if multiple calls try and access the file when the first call deletes it!  
</code></pre>

<p>WARN  11:05:27,577 RMDTrackBuilder - Index file /home/pipeman/gatk_resources/Mills_and_1000G_gold_standard.indels.b37.vcf.idx is out of date (index older than input file), deleting and updating the index file 
INFO  11:05:31,699 RMDTrackBuilder - Writing Tribble index to disk for file /home/pipeman/gatk_resources/Mills_and_1000G_gold_standard.indels.b37.vcf.idx 
```</p>
<h2 id="cnv-tools-to-think-about">CNV tools to think about</h2>
<p>EXCAVATOR: detecting copy number variants from whole-exome sequencing data @ http://genomebiology.com/2013/14/10/R120</p>
<blockquote>
<p>We developed a novel software tool, EXCAVATOR, for the detection of copy number variants (CNVs) from whole-exome sequencing data. EXCAVATOR combines a three-step normalization procedure with a novel heterogeneous hidden Markov model algorithm and a calling method that classifies genomic regions into five copy number states. We validate EXCAVATOR on three datasets and compare the results with three other methods. These analyses show that EXCAVATOR outperforms the other methods and is therefore a valuable tool for the investigation of CNVs in largescale projects, as well as in clinical research and diagnostics. EXCAVATOR is freely available at http://sourceforge.net/projects/excavatortool/ webcite.</p>
</blockquote>
<hr />
<h2 id="useful-links">Useful Links</h2>
<ul>
<li>https://bcbio.wordpress.com/  </li>
<li>https://basecallbio.wordpress.com/2013/04/23/base-quality-score-rebinning/  </li>
<li>https://github.com/statgen/bamUtil  </li>
<li>http://genome.sph.umich.edu/wiki/BamUtil:_recab  </li>
<li>https://github.com/chapmanb/bcbio.variation  </li>
<li>http://plagnol-lab.blogspot.co.uk/2013/11/faq-and-clarifications-for-exomedepth-r.html</li>
</ul>
</div>
            
        </div>

        <footer class="col-md-12">
            <hr>
            
            <p>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p>
        </footer>

        

        <script src="https://code.jquery.com/jquery-1.10.2.min.js"></script>
        <script src="./js/bootstrap-3.0.3.min.js"></script>
        <script src="./js/prettify-1.0.min.js"></script>
        <script src="./js/highlight.pack.js"></script>
        <script src="./js/base.js"></script>
    </body>
</html>