<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Ngseasy : Dockerised Next Generation Sequencing Pipeline (QC, Align, Calling, Annotation)">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Ngseasy</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/KHP-Informatics/ngseasy">View on GitHub</a>

          <h1 id="project_title">Ngseasy</h1>
          <h2 id="project_tagline">Dockerised Next Generation Sequencing Pipeline (QC, Align, Calling, Annotation)</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/KHP-Informatics/ngseasy/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/KHP-Informatics/ngseasy/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <p><img src="https://github.com/KHP-Informatics/ngseasy/blob/sjn_dev/figs/NGSeasy_logo_0.0.1.png" alt="NGSeasy_logo" title="NGSeasy_logo"></p>

<h1>
<a id="ngseasy-beta" class="anchor" href="#ngseasy-beta" aria-hidden="true"><span class="octicon octicon-link"></span></a>NGSeasy (beta)</h1>

<p>** This is the latest dev project **<br>
<strong>note</strong>: undergoing massive re-dev , many links broken...stay tuned and email us. give us a few weeks.  </p>

<p><a href="http://core.brc.iop.kcl.ac.uk">Funded by Biomedical Research Centre</a>: <a href="http://core.brc.iop.kcl.ac.uk">http://core.brc.iop.kcl.ac.uk</a></p>

<p>Publication: pending</p>

<p>Authors: Stephen J Newhouse, Amos Folarin , Maximilian Kerz<br>
Release Version: <strong>1.0.0</strong>  </p>

<hr>

<h3>
<a id="a-dockerized-ngs-pipeline-and-tool-box" class="anchor" href="#a-dockerized-ngs-pipeline-and-tool-box" aria-hidden="true"><span class="octicon octicon-link"></span></a>[A <a href="https://www.docker.com/">Dockerized</a> NGS pipeline and tool-box]</h3>

<hr>

<p><a href="https://twitter.com/share">Tweet</a></p>



<p><strong>With NGSeasy you can now have full suite of NGS tools up and running on any high end workstation in an afternoon</strong></p>

<p><strong>Note:</strong> NGSeasy is under <strong>heavy development</strong> and the code and docs evolve quickly.  </p>

<ul>
<li><p><strong>NGSeasy-v1.0 Full Production release will be available Early 2015</strong>  </p></li>
<li><p><strong>NGSeasy-v1.0.0b Full Production release contains most of the core fucntionality to go from raw fastq to raw vcf calls</strong></p></li>
<li><strong>NGSeasy updates every 12 months</strong></li>
<li><strong>GUI in development</strong></li>
<li>
<strong>Contact us of Reference Genomes and resource files</strong><br>
</li>
</ul>

<hr>

<blockquote>
<p>NGSeasy is completely open source and we encourage interested folks to jump in and get involved in the dev with us.</p>
</blockquote>

<hr>

<h1>
<a id="ngseasy-easy-analysis-of-next-generation-sequencing" class="anchor" href="#ngseasy-easy-analysis-of-next-generation-sequencing" aria-hidden="true"><span class="octicon octicon-link"></span></a>NGSeasy (Easy Analysis of Next Generation Sequencing)</h1>

<p>We present <strong>NGSeasy (Easy Analysis of Next Generation Sequencing)</strong>, a flexible and easy-to-use NGS pipeline for automated alignment, quality control, variant calling and annotation. The pipeline allows users with minimal computational/bioinformatic skills to set up and run an NGS analysis on their own samples, in less than an afternoon, on any operating system (Windows, iOS or Linux) or infrastructure (workstation, cluster or cloud).</p>

<p>NGS pipelines typically utilize a large and varied range of software components and incur a substantial configuration burden during deployment which limits their portability to different computational environments. NGSeasy simplifies this by providing the pipeline components encapsulated in Dockerâ„¢ containers and bundles in a wide choice of tools for each module. Each module of the pipeline represents one functional grouping of tools (e.g. sequence alignment, variant calling etc.).</p>

<p>Deploying the pipeline is as simple as pulling the container images from the public repository into any host running Docker. NGSeasy can be deployed on any medium to high-end workstation, high performance computer cluster and compute clouds (public/private cloud computing) - enabling instant access to elastic scalability without investment overheads for additional compute hardware and makes open and reproducible research straight forward for the greater scientific community.</p>

<h3>
<a id="advantages" class="anchor" href="#advantages" aria-hidden="true"><span class="octicon octicon-link"></span></a>Advantages</h3>

<ul>
<li>Easy to use for non-informaticians.<br>
</li>
<li>All run from a single config file that can be made in Excel.<br>
</li>
<li>User can select from mutiple aligners, variant callers and variant annotators</li>
<li>No scary python, .yaml or .json files...just one simple Excel workbook saved as a textfile.<br>
</li>
<li>Just follow our simple set of instructions and NGS away!<br>
</li>
<li>Choice of aligners and variant callers and anntators<br>
</li>
<li>Allows reproducible research<br>
</li>
<li>Version controlled for auditing<br>
</li>
<li>Customisable<br>
</li>
<li>Easy to add new tools<br>
</li>
<li>If it's broke...we will fix it..</li>
<li>Enforced naming convention and directory structures<br>
</li>
<li>Allows users to run "Bake Offs" between tools with ease<br>
</li>
</ul>

<p>We have adapted the current best practices from the Genome Analysis Toolkit (GATK, <a href="http://www.broadinstitute.org/gatk/guide/best-practices">http://www.broadinstitute.org/gatk/guide/best-practices</a>)  for processing raw alignments in SAM/BAM format and variant calling. The current workflow, has been optimised for Illumina platforms, but can easily be adapted for other sequencing platforms, with minimal effort.  </p>

<p>As the containers themselves can be run as executables with pre-specified cpu and RAM resources, the orchestration of the pipeline can be placed under the control of conventional load balancers if this mode is required.  </p>

<hr>

<h3>
<a id="author-contact-details" class="anchor" href="#author-contact-details" aria-hidden="true"><span class="octicon octicon-link"></span></a>Author Contact Details</h3>

<p>Please contact us for help/guidance on using the beta release. </p>

<ul>
<li><p>Amos Folarin <a href="mailto:amosfolarin@gmail.com">amosfolarin@gmail.com</a>  <a href="https://twitter.com/amosfolarin?lang=en">@amosfolarin</a><br>
<a href="http://www.linkedin.com/pub/amos-folarin/34/b06/978">
<img src="http://www.linkedin.com/img/webpromo/btn_viewmy_160x33.png" width="160" height="33" alt="View Amos's profile on LinkedIn">
</a></p></li>
<li><p>Stephen J Newhouse <a href="mailto:stephen.j.newhouse@gmail.com">stephen.j.newhouse@gmail.com</a> <a href="https://twitter.com/s_j_newhouse?lang=en">@s_j_newhouse</a><br>
<a href="http://uk.linkedin.com/pub/dr-stephen-newhouse/29/89a/11a">
<img src="http://www.linkedin.com/img/webpromo/btn_viewmy_160x33.png" width="160" height="33" alt="View Steve's profile on LinkedIn">
</a></p></li>
</ul>

<p><strong>Lets us know if you want other tools added to NGSeasy</strong></p>

<p><em>Institution: NIHR Maudsley Biomedical Research Centre For Mental Health and Dementia Unit (Denmark Hill), at The Institute of Psychiatry, Psychology &amp; Neuroscience (IoPPN), Kings College London</em> </p>

<hr>

<h1>
<a id="overview-of-the-ngseasy-pipeline-components" class="anchor" href="#overview-of-the-ngseasy-pipeline-components" aria-hidden="true"><span class="octicon octicon-link"></span></a>Overview of the NGSeasy Pipeline Components</h1>

<p>The basic pipeline contains all the basic tools needed for manipulation and 
quality control of raw fastq files (ILLUMINA focused), SAM/BAM manipulation,
alignment, cleaning (based on GATK best practises [<a href="http://www.broadinstitute.org/gatk/guide/best-practices">http://www.broadinstitute.org/gatk/guide/best-practices</a>]) and first pass
variant discovery. Separate containers are provided for indepth variant annotation,
structural variant calling, basic reporting and visualisations.  </p>

<p><img src="https://github.com/KHP-Informatics/ngs/blob/dev2/figs/ngsEASY_component_visualisation.png" alt="ngsEASY" title="Dockerized NGS Pipeline"></p>

<hr>

<h1>
<a id="the-full-ngseasy-pipeline" class="anchor" href="#the-full-ngseasy-pipeline" aria-hidden="true"><span class="octicon octicon-link"></span></a>The Full NGSeasy pipeline</h1>

<p>The NGSeasy pipelines implement the following :-   </p>

<ul>
<li><p><strong>Quality control of raw fastq</strong> files using <strong><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FASTQC</a></strong>  </p></li>
<li><p><strong>Read trimming</strong> using <strong><a href="http://www.usadellab.org/cms/?page=trimmomatic">TRIMMOMATIC</a></strong>.   </p></li>
<li>
<p><strong>Alignment</strong> using one of </p>

<ul>
<li>
<strong><a href="http://bio-bwa.sourceforge.net/">BWA</a></strong><br>
</li>
<li>
<strong><a href="http://www.well.ox.ac.uk/project-stampy">STAMPY</a></strong><br>
</li>
<li>
<strong><a href="http://www.novocraft.com">NOVOALIGN</a></strong><br>
</li>
<li>
<strong><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">BOWTIE2</a></strong><br>
</li>
<li>
<strong><a href="http://snap.cs.berkeley.edu/">SNAP</a></strong><br>
</li>
</ul>
</li>
<li><p><strong>SAM/BAM sorting and indexing</strong> with <strong><a href="https://github.com/lomereiter/sambamba">SAMBAMBA</a></strong>.  </p></li>
<li><p><strong>Read Group information added</strong> using <strong><a href="http://broadinstitute.github.io/picard/">PICARDTOOLS</a>:<a href="http://broadinstitute.github.io/picard/command-line-overview.html#AddOrReplaceReadGroups">AddOrReplaceReadGroups</a></strong> </p></li>
<li><p><strong>Duplicate marking</strong> with <strong><a href="https://github.com/GregoryFaust/samblaster">SAMBLASTER</a></strong>.  </p></li>
</ul>

<blockquote>
<p>For academic users and/or commercial/clinical groups whom have paid for GATK licensing, the next steps are to perform   </p>
</blockquote>

<ul>
<li>
<strong>Indel indel realignment and base quality score recalibration</strong> using <strong><a href="https://www.broadinstitute.org/gatk/">GATK</a></strong> built in tools :

<ul>
<li>
<strong><a href="https://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_gatk_tools_walkers_indels_RealignerTargetCreator.php">GATK:RealignerTargetCreator</a></strong> </li>
<li>
<strong><a href="https://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_gatk_tools_walkers_indels_IndelRealigner.php">GATK:IndelRealigner</a></strong> </li>
<li>
<strong><a href="https://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_gatk_tools_walkers_bqsr_BaseRecalibrator.php">GATK:BaseRecalibrator</a></strong> </li>
</ul>
</li>
</ul>

<blockquote>
<p>For the non-GATK version    </p>
</blockquote>

<ul>
<li>
<p><strong>Base quality score recalibration</strong> using <a href="http://genome.sph.umich.edu/wiki/BamUtil">BamUtil</a>  </p>

<ul>
<li>
<strong><a href="http://genome.sph.umich.edu/wiki/BamUtil:_recab">BamUtil:recab</a></strong> </li>
</ul>
</li>
<li>
<p><strong>Post alignment quality control and reporting</strong> is performed usng a number of tools and custom scripts: </p>

<ul>
<li><strong><a href="https://github.com/samtools/samtools">SAMTOOLS:flagstats</a></strong></li>
<li><strong><a href="https://github.com/arq5x/bedtools2">BEDTOOLS:genomecov</a></strong></li>
<li><strong><a href="https://github.com/arq5x/bedtools2">BEDTOOLS:bamtobed</a></strong></li>
<li>
<strong><a href="http://broadinstitute.github.io/picard/command-line-overview.html#CollectMultipleMetrics">PICARDTOOLS:CollectMultipleMetrics</a></strong><br>
</li>
<li>
<strong><a href="http://broadinstitute.github.io/picard/command-line-overview.html#CollectAlignmentSummaryMetrics">PICARDTOOLS:CollectAlignmentSummaryMetrics</a></strong><br>
</li>
<li>
<strong><a href="http://broadinstitute.github.io/picard/command-line-overview.html#CollectWgsMetrics">PICARDTOOLS:CollectWgsMetrics</a></strong><br>
</li>
<li>
<strong><a href="http://broadinstitute.github.io/picard/command-line-overview.html#CollectTargetedPcrMetrics">PICARDTOOLS:CollectTargetedPcrMetrics</a></strong> (coming soon)<br>
</li>
</ul>
</li>
<li>
<p><strong>SNP and small INDEL</strong> calling using one of the following or a combibation of these tools, if the <code>ensemble</code> method is called using <strong><a href="https://github.com/chapmanb/bcbio.variation">bcbio.variation variant-ensemble</a></strong></p>

<ul>
<li>
<strong><a href="https://github.com/ekg/freebayes">FREEBAYES</a></strong> </li>
<li>
<strong><a href="http://www.well.ox.ac.uk/platypus">PLATYPUS</a></strong> </li>
<li>
<strong><a href="https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_genotyper_UnifiedGenotyper.php">GATK:UnifiedGenotyper</a></strong> </li>
<li>
<strong><a href="https://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_gatk_tools_walkers_haplotypecaller_HaplotypeCaller.php">GATK:HaplotypeCaller</a></strong> </li>
</ul>
</li>
<li>
<p><strong>Structural Variant (CNV)</strong> calling using one of the following or or a combibation of if the <code>ensemble</code> methods are called:- </p>

<ul>
<li>
<strong><a href="https://github.com/tobiasrausch/delly">DELLY</a> : in Dev</strong> </li>
<li><strong><a href="https://github.com/arq5x/lumpy-sv/">LUMPY</a>: in Dev</strong></li>
<li><strong><a href="http://www.bioinf.jku.at/software/cnmops/">cn.MOPS</a>: in Dev</strong></li>
<li><strong><a href="https://www.stt.msu.edu/users/hengwang/mHMM.html">m-HMM</a>: in Dev</strong></li>
<li><strong><a href="http://cran.r-project.org/web/packages/ExomeDepth/index.html">ExomeDepth</a>: in Dev</strong></li>
<li>
<strong><a href="http://www-genepi.med.utah.edu/suppl/SLOPE/index.html">SLOPE</a>: in Dev and not tested</strong><br>
</li>
<li>
<strong><a href="http://sv.gersteinlab.org/">cnvnator</a>: in Dev</strong><br>
</li>
</ul>
</li>
<li>
<p><strong>Variant annotation</strong> using using one of the following or a combibation of if the <code>ensemble</code> methods are called. </p>

<ul>
<li>
<strong><a href="http://snpeff.sourceforge.net/">SnpEff</a>: in Dev</strong> </li>
<li>
<strong><a href="http://www.openbioinformatics.org/annovar/">ANNOVAR</a>: in Dev</strong> </li>
<li><strong><a href="http://www.ensembl.org/info/docs/tools/vep/index.html">VEP</a>: in Dev</strong></li>
</ul>
</li>
<li><p><strong>Variant reporting</strong> using custom scripts</p></li>
</ul>

<p><strong>Note</strong> Some of the later functions i.e. variant annotation and qc reporting are still in dev.  </p>

<hr>

<p><strong>We highly recommed read trimming prior to alignment.</strong>
We have noticed considerable speed-ups in alignmnet time and increased quality of SNP/INDEL calls using trimmed vs raw fastq.  </p>

<p><strong>Base quality score recalibration is also recommended.</strong><br>
As an alternative to GATK, we have added fucntionality for use of 
<strong><a href="https://github.com/statgen/bamUtil">BamUtil</a>:<a href="http://genome.sph.umich.edu/wiki/BamUtil:_recab">recab</a></strong> 
for base quality score recalibration.</p>

<p><strong>Non-GATK users</strong> </p>

<ul>
<li>are encouraged to use aligners such as <strong><a href="http://www.well.ox.ac.uk/project-stampy">stampy</a></strong> and <strong><a href="http://www.novocraft.com">novoalign</a></strong> that perform base quality score recal on the fly.<br>
</li>
<li>are encouraged to use variant callers that perform local re-aligmnet around candidate sites to mitigate the need for the indel realignment stages.<br>

<ul>
<li><strong><a href="https://github.com/ekg/freebayes">freebayes</a></strong></li>
<li><strong><a href="http://www.well.ox.ac.uk/platypus">platypus</a></strong></li>
</ul>
</li>
</ul>

<hr>

<h3>
<a id="coming-soon" class="anchor" href="#coming-soon" aria-hidden="true"><span class="octicon octicon-link"></span></a>Coming Soon</h3>

<ul>
<li>New Aligners:- <a href="http://snap.cs.berkeley.edu/">SNAP</a>, GSNAP, mr- and mrs-Fast,gem</li>
<li><a href="https://github.com/amplab/snap">https://github.com/amplab/snap</a></li>
<li>SLOPE (CNV fo targetted NSG) </li>
<li>Cancer Pipelines</li>
<li>Annotation Pipelines and Databases</li>
<li>Visualisation Pipelines</li>
<li>Var Callers:- VarScan2</li>
<li>SGE scripts and basic BASH scrips for running outside of Docker</li>
<li>biobambam <a href="https://github.com/gt1/biobambam">https://github.com/gt1/biobambam</a><br>
</li>
<li>bamaddrg <a href="https://github.com/ekg/bamaddrg">https://github.com/ekg/bamaddrg</a><br>
</li>
<li>bamtools <a href="https://github.com/ekg/bamtools">https://github.com/ekg/bamtools</a><br>
</li>
</ul>

<hr>

<h2>
<a id="a-special-note-on-the-base-image" class="anchor" href="#a-special-note-on-the-base-image" aria-hidden="true"><span class="octicon octicon-link"></span></a>A Special note on the base image.</h2>

<p>We include the following - what we think of as - <strong><em>NGS Powertools</em></strong> in the <strong><a href="">compbio/ngseasy-base</a></strong> image. 
These are all tools that allow the user to slice and dice BED/SAM/BAM/VCF files in multiple ways.</p>

<ol>
<li> samtools<br>
</li>
<li> bcftools<br>
</li>
<li> vcftools<br>
</li>
<li> vcflib<br>
</li>
<li> bamUtil<br>
</li>
<li> bedtools2<br>
</li>
<li> ogap<br>
</li>
<li> samblaster<br>
</li>
<li> sambamba<br>
</li>
<li>bamleftalign<br>
</li>
<li>seqtk<br>
</li>
<li>parallel<br>
</li>
</ol>

<p>This image is used as the base of all our compbio/ngseasy-* tools.   </p>

<p><strong>Why not a separate containers per application?</strong> The more docker-esque approach, would be to have separate containers for each NGS tool. 
However, this belies the fact that many of these tools interact in a deep way, allowing pipes and streamlined system calls for manipulating
the output of NGS pipelines (BED/SAM/BAM/VCF files). Therefore, we built  these into a single development environment for ngseasy.</p>

<hr>

<h1>
<a id="dockerised-ngseasy" class="anchor" href="#dockerised-ngseasy" aria-hidden="true"><span class="octicon octicon-link"></span></a>Dockerised NGSeasy</h1>

<p><img src="https://github.com/KHP-Informatics/ngs/blob/master/figs/Docker_container_engine_logo.png" alt="docker" title="Docker">  </p>

<p>The following section describes getting the Dockerised NGSeasy Pipeline(s) and Resources, project set up and running NGSeasy.</p>

<p>Getting all resources and building required tools will take a few hours depending on network connections and any random "ghosts in the machine" - half a day in reality.
But once you're set up, thats it - you are good to go.</p>

<hr>

<h2>
<a id="1-install-docker" class="anchor" href="#1-install-docker" aria-hidden="true"><span class="octicon octicon-link"></span></a>1. Install Docker</h2>

<p>Follow the simple instructions in the links provided below  </p>

<ul>
<li>
<a href="https://docs.docker.com/installation/mac/">Mac</a><br>
</li>
<li><a href="https://docs.docker.com/installation/windows/">Windows</a></li>
<li><a href="https://docs.docker.com/installation/ubuntulinux/">Ubuntu</a></li>
</ul>

<p>A full set of instructions for multiple operating systems are available on the <a href="https://docs.docker.com/installation/">Docker website</a>.</p>

<h2>
<a id="2-get-ngseasy" class="anchor" href="#2-get-ngseasy" aria-hidden="true"><span class="octicon octicon-link"></span></a>2. Get NGSeasy</h2>

<p>We provide a simple Makefile to pull all of the public nsgeasy components, scripts and set up to correct project directory structre on your local machines.</p>

<div class="highlight highlight-bash"><pre>
git clone https://github.com/KHP-Informatics/ngseasy.git

<span class="pl-s3">cd</span> ngseasy

make all
</pre></div>

<p>Setting up the initial project can take up a day, depending on your local network connections and speeds.</p>

<h2>
<a id="3-set-up-ngseasy-project-configuration-file" class="anchor" href="#3-set-up-ngseasy-project-configuration-file" aria-hidden="true"><span class="octicon octicon-link"></span></a>3. Set up NGSeasy Project configuration file</h2>

<p>In Excel make config file and save as [TAB] Delimited file with <code>.tsv</code> extenstion.<br>
See Example provided and <a href="https://docs.google.com/spreadsheets/d/1kp1Nyw0x3zXqO2Wm2Z25ErJ0Z-Uoab8tjRPq9h4sonk/edit?usp=sharing">GoogleDoc</a>. Remove the header from this file before running the pipeline. This sets up Information related to: Project Name, Sample Name, Library Type, Pipeline to call, NCPU.</p>

<p>The [config.file.tsv] should contain the following 15 columns for each sample to be run through a pipeline:- </p>

<table>
<thead>
<tr>
<th>Variable</th>
<th>type</th>
<th>Description</th>
<th>Options/Examples</th>
</tr>
</thead>
<tbody>
<tr>
<td>POJECT_ID</td>
<td>string</td>
<td>Project ID</td>
<td>Cancer</td>
</tr>
<tr>
<td>SAMPLE_ID</td>
<td>string</td>
<td>Sample ID</td>
<td>T100</td>
</tr>
<tr>
<td>FASTQ1</td>
<td>string</td>
<td>Raw fastq file name read 1</td>
<td>foo_1_fq.gz</td>
</tr>
<tr>
<td>FASTQ2</td>
<td>string</td>
<td>Raw fastq file name read 1</td>
<td>foo_2_fq.gz</td>
</tr>
<tr>
<td>PROJECT_DIR</td>
<td>string</td>
<td>Project Directory</td>
<td>/medida/ngs_projects</td>
</tr>
<tr>
<td>DNA_PREP_LIBRARY_ID</td>
<td>string</td>
<td>DNA Libray Prep ID</td>
<td>Custom_Cancer</td>
</tr>
<tr>
<td>NGS_PLATFORM</td>
<td>string</td>
<td>Platform Name</td>
<td>ILLUMINA</td>
</tr>
<tr>
<td>NGS_TYPE</td>
<td>string</td>
<td>Experiment type</td>
<td>WGS, WEX, TGS</td>
</tr>
<tr>
<td>BAIT</td>
<td>string</td>
<td>user supplied bed file</td>
<td></td>
</tr>
<tr>
<td>CAPTURE</td>
<td>string</td>
<td>user supplied bed file</td>
<td></td>
</tr>
<tr>
<td>FASTQC</td>
<td>string</td>
<td>run FastQc</td>
<td>skip, qc-fastq</td>
</tr>
<tr>
<td>TRIM</td>
<td>string</td>
<td>run Trimmomatic</td>
<td>skip, qc-trimm, qc-adaptor</td>
</tr>
<tr>
<td>BSQR</td>
<td>string</td>
<td>Base Quality Score Recalibration</td>
<td>skip, bam-recab, gatk-recab</td>
</tr>
<tr>
<td>REALN</td>
<td>string</td>
<td>Bam Realignment around indels</td>
<td>skip,bam-realn, gatk-realn</td>
</tr>
<tr>
<td>ALIGNER</td>
<td>string</td>
<td>Aligner</td>
<td>skip, bwa, bowtie2, stampy, snap, novoalign</td>
</tr>
<tr>
<td>VARCALLER</td>
<td>string</td>
<td>Variant Caller</td>
<td>ensemble,ensemble-fast, freebayes, platypus, UnifiedGenotyper, HaplotypeCaller</td>
</tr>
<tr>
<td>CNV</td>
<td>string</td>
<td>CNV Caller</td>
<td>skip, lump, delly, exomedepth</td>
</tr>
<tr>
<td>ANNOTATOR</td>
<td>string</td>
<td>Choose annotator</td>
<td>skip</td>
</tr>
<tr>
<td>CLEANUP</td>
<td>string</td>
<td>Clean Up Files (TRUE/FALSE)</td>
<td>TRUE/FALSE</td>
</tr>
<tr>
<td>NCPU</td>
<td>number</td>
<td>Number of cores to call</td>
<td>1..n</td>
</tr>
<tr>
<td>VERSION</td>
<td>number</td>
<td>NGSeasy Version</td>
<td>1.0</td>
</tr>
<tr>
<td>NGSUSER</td>
<td>string</td>
<td>user email address</td>
<td><a href="mailto:stephen.j.newhouse@gmail.com">stephen.j.newhouse@gmail.com</a></td>
</tr>
</tbody>
</table>

<h2>
<a id="4-run-ngseasy" class="anchor" href="#4-run-ngseasy" aria-hidden="true"><span class="octicon octicon-link"></span></a>4. Run NGSeasy</h2>

<div class="highlight highlight-bash"><pre></pre></div>

<p>All NGSeasy Docker images can be pulled down from <strong><a href="https://hub.docker.com/u/compbio/">compbio Docker Hub</a></strong> or using the Makefile.<br>
We provide an Amazon EBS data volume with indexed genomes: XXXXXX  </p>

<hr>

<h3>
<a id="dockerised-ngs-tools" class="anchor" href="#dockerised-ngs-tools" aria-hidden="true"><span class="octicon octicon-link"></span></a>Dockerised NGS Tools</h3>

<p>The following opensource tools are all provided. </p>

<table>
<thead>
<tr>
<th>Tool</th>
<th>Build</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-base/">ngseasy-base</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-fastqc">fastqc</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-trimmomatic">trimmomatic</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-bwa">bwa</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-bowtie">bowtie</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-picardtools">picardtools</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-samtools">samtools</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-freebayes/">freebayes</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-bedtools/">bedtools</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-bcbiovar/">bcbiovar</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-delly">delly</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-lumpy">lumpy</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-cnmops">cnmops</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-mhmm">mhmm</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-exomedepth">exomedepth</a></td>
<td>automated build</td>
</tr>
<tr>
<td><a href="https://registry.hub.docker.com/u/compbio/ngseasy-bamutil">bamutil</a></td>
<td>automated build</td>
</tr>
</tbody>
</table>

<p>samtools includes bcftools and htslib  </p>

<p>Its as easy as: - </p>

<div class="highlight highlight-bash"><pre>docker pull compbio/ngseasy-<span class="pl-vo">${TOOL}</span></pre></div>

<hr>

<h2>
<a id="the-ngseasy-project-directory" class="anchor" href="#the-ngseasy-project-directory" aria-hidden="true"><span class="octicon octicon-link"></span></a>The NGSeasy project directory</h2>

<p>The user needs to make the relevent directory structures on their local machine before starting an NGS run. </p>

<p>On our sysetm we typically set up a top-level driectory called <strong>ngs_projects</strong> within which we store output from all our individual NGS projects. </p>

<p>Here we are working from local top level directory called <strong>media/</strong>, but this can really be any folder on your local system ie your home directory <strong>~/${USER}</strong>.  </p>

<p>Within this directory <strong>media</strong> we make the following folders: - </p>

<div class="highlight highlight-bash"><pre>ngs_projects  
<span class="pl-k">|</span>  
<span class="pl-k">|</span>__raw_fastq  
<span class="pl-k">|</span>__config_files  
<span class="pl-k">|</span>__reference_genomes_b37  
<span class="pl-k">|</span>__gatk_resources  
<span class="pl-k">|</span>__ngseasy</pre></div>

<p><strong>Note</strong> The following directories are obtained in step <strong><a href="https://github.com/KHP-Informatics/ngs/tree/master#4-download-ngseasy-resources">4. Download NGSeasy Resources</a></strong>.<br>
*<em>- reference_genomes_b37  *</em><br>
*<em>- gatk_resources *</em>  </p>

<p><strong>Move to media</strong></p>

<div class="highlight highlight-bash"><pre><span class="pl-c"># Move to media/</span>
<span class="pl-s3">cd</span> media</pre></div>

<p><strong>make toplevel ngs_projects folder</strong></p>

<div class="highlight highlight-bash"><pre><span class="pl-c"># make toplevel NGS folder</span>
mkdir ngs_projects </pre></div>

<p><strong>make fast_raw folder</strong></p>

<div class="highlight highlight-bash"><pre><span class="pl-c"># fastq staging area</span>
mkdir ngs_projects/fastq_raw </pre></div>

<p><strong>make config_files folder</strong></p>

<div class="highlight highlight-bash"><pre><span class="pl-c"># config files</span>
mkdir ngs_projects/config_files </pre></div>

<p><strong>make ngseasy folder</strong></p>

<div class="highlight highlight-bash"><pre><span class="pl-c"># NGSeasy scripts</span>
mkdir ngs_projects/ngseasy </pre></div>

<hr>

<h2>
<a id="4-download-ngseasy-resources" class="anchor" href="#4-download-ngseasy-resources" aria-hidden="true"><span class="octicon octicon-link"></span></a>4. Download NGSeasy Resources</h2>

<p>Download the indexed reference genomes and example data for use with NGSeasy.</p>

<p><strong>NGSeasy Resources:-</strong>  </p>

<ul>
<li>
<strong>reference_genomes_b37.tgz</strong> b37 reference genomes indexed for use with all provided aligners (BWA, Bowtie2, Stampy, Novoalign) and annotation bed files for use with pipeline scripts</li>
<li>
<strong>gatk_resources.tar.gz</strong> gatk resources bundle</li>
<li>
<strong>fastq_example.tgz</strong> Example 75bp PE Illumina Whole Exome Sequence fastq data for <strong>NA12878</strong>
</li>
<li>Annotation Databases Coming in the next update </li>
</ul>

<p><strong>Download the data to the top level directory</strong></p>

<h3>
<a id="ftp-details" class="anchor" href="#ftp-details" aria-hidden="true"><span class="octicon octicon-link"></span></a>FTP Details</h3>

<ul>
<li>
<strong>ftp:</strong>  159.92.120.21<br>
</li>
<li>
<strong>user:</strong> compbio-public<br>
</li>
<li>
<strong>pwd:</strong>  compbio-public<br>
</li>
<li>
<strong>port:</strong> 21<br>
</li>
</ul>

<p><strong>Move to top level directory</strong></p>

<div class="highlight highlight-bash"><pre><span class="pl-s3">cd</span> ngs_projects</pre></div>

<p><strong>FTP NGSeasy Resources</strong></p>

<div class="highlight highlight-bash"><pre>ftp 159.92.120.21</pre></div>

<p><strong>mget NGSeasy Resources</strong></p>

<div class="highlight highlight-bash"><pre>ftp<span class="pl-k">&gt;</span> <span class="pl-s3">cd</span> /Public/NGSeasy_Public_Resources
ftp<span class="pl-k">&gt;</span> prompt off
ftp<span class="pl-k">&gt;</span> mget <span class="pl-k">*</span>.gz
ftp<span class="pl-k">&gt;</span> <span class="pl-s3">exit</span></pre></div>

<p>I would recommend using a separate program like <a href="https://filezilla-project.org/">FileZilla</a>, which will make it much easier for you to set up and manage your file transfers</p>

<p><strong>Extract NGSeasy Resources</strong></p>

<div class="highlight highlight-bash"><pre><span class="pl-c"># Extract resources</span>
<span class="pl-s3">cd</span> ngs_projects/
tar xvf gatk_resources.tgz<span class="pl-k">;</span> 
<span class="pl-s3">cd</span> ngs_projects/gatk_resources
gunzip <span class="pl-k">*</span></pre></div>

<p><strong>Extract NGSeasy Reference Genomes</strong></p>

<div class="highlight highlight-bash"><pre><span class="pl-c"># Extract Reference Genomes</span>
<span class="pl-s3">cd</span> ngs_projects/
tar xvf reference_genomes_b37.tgz<span class="pl-k">;</span> 
<span class="pl-s3">cd</span> ngs_projects/reference_genomes_b37
gunzip <span class="pl-k">*</span></pre></div>

<hr>

<h3>
<a id="gatk-resources" class="anchor" href="#gatk-resources" aria-hidden="true"><span class="octicon octicon-link"></span></a>GATK Resources</h3>

<ul>
<li>
<a href="https://www.broadinstitute.org/gatk/guide/article.php?id=1215">https://www.broadinstitute.org/gatk/guide/article.php?id=1215</a><br>
</li>
<li>
<a href="https://www.broadinstitute.org/gatk/guide/article.php?id=1213">https://www.broadinstitute.org/gatk/guide/article.php?id=1213</a><br>
</li>
</ul>

<p><strong>Downloading</strong></p>

<pre><code>location: ftp.broadinstitute.org
username: gsapubftp-anonymous
password: &lt;blank&gt;
</code></pre>

<p><strong>b37 Resources: the Standard Data Set</strong></p>

<ul>
<li>Reference sequence (standard 1000 Genomes fasta) along with fai and dict files</li>
<li>dbSNP in VCF. This includes two files:

<ul>
<li>The most recent dbSNP release
This file subsetted to only sites discovered in or before dbSNPBuildID 129, which excludes the impact of the 1000 Genomes project and is useful for evaluation of dbSNP rate and Ti/Tv values at novel sites.</li>
<li>HapMap genotypes and sites VCFs</li>
</ul>
</li>
<li>OMNI 2.5 genotypes for 1000 Genomes samples, as well as sites, VCF
The current best set of known indels to be used for local realignment (note that we don't use dbSNP for this anymore); use both files:</li>
<li>1000G_phase1.indels.b37.vcf (currently from the 1000 Genomes Phase I indel calls)</li>
<li>Mills_and_1000G_gold_standard.indels.b37.sites.vcf</li>
<li>A large-scale standard single sample BAM file for testing:

<ul>
<li>NA12878.HiSeq.WGS.bwa.cleaned.recal.hg19.20.bam containing ~64x reads of NA12878 on chromosome 20
The results of the latest UnifiedGenotyper with default arguments run on this data set (NA12878.HiSeq.WGS.bwa.cleaned.recal.hg19.20.vcf)</li>
</ul>
</li>
</ul>

<hr>

<h2>
<a id="5-get-ngseasy-sripts" class="anchor" href="#5-get-ngseasy-sripts" aria-hidden="true"><span class="octicon octicon-link"></span></a>5. Get NGSeasy Sripts</h2>

<p>We then need to get the latest NGSeasy scripts from <a href="https://github.com/KHP-Informatics/ngs">GitHub</a> . The user is required to download the scripts to the <code>ngseasy</code> directory</p>

<p><strong>move to the <code>ngseasy</code> directory</strong></p>

<div class="highlight highlight-bash"><pre><span class="pl-s3">cd</span> /media/ngs_projects/nsgeasy</pre></div>

<p><strong>clone the <a href="https://github.com/KHP-Informatics/ngs">ngs</a> repository</strong></p>

<div class="highlight highlight-bash"><pre>git clone https://github.com/KHP-Informatics/ngs.git</pre></div>

<p><strong>add <code>nsgeasy/ngs/bin</code> to your system PATH</strong></p>

<div class="highlight highlight-bash"><pre><span class="pl-s">export</span> PATH=<span class="pl-vo">$PATH</span>:/media/ngs_projects/nsgeasy/ngs/bin</pre></div>

<p><strong>or add to global .bashrc</strong></p>

<div class="highlight highlight-bash"><pre><span class="pl-s3">echo</span> <span class="pl-s1"><span class="pl-pds">"</span>export PATH=<span class="pl-vo">$PATH</span>:/media/ngs_projects/nsgeasy/ngs/bin<span class="pl-pds">"</span></span> <span class="pl-k">~</span>/.bashrc
<span class="pl-s3">source</span> <span class="pl-k">~</span>/.bashrc</pre></div>

<p><strong>alternatively donwload the scripts from our <a href="https://github.com/KHP-Informatics/ngs">GitHub Release</a></strong> </p>

<hr>

<h2>
<a id="6-manually-build-required-ngseasy-container-images" class="anchor" href="#6-manually-build-required-ngseasy-container-images" aria-hidden="true"><span class="octicon octicon-link"></span></a>6. Manually Build required NGSeasy Container Images</h2>

<p>Currently we are not able to automatically build some of the tools in pre-built docker containers due to licensing restrictions. </p>

<p>Some of the software has restrictions on use particularly for commercial 
purposes. Therefore if you wish to use this for commercial purposes, then you 
leagally have to approach the owners of the various components yourself!  </p>

<p><strong>Software composing the pipeline requiring registration:-</strong>  </p>

<ul>
<li>novoalign <a href="http://www.novocraft.com/">http://www.novocraft.com/</a><br>
</li>
<li>Stampy <a href="http://www.well.ox.ac.uk/project-stampy">http://www.well.ox.ac.uk/project-stampy</a><br>
</li>
<li>Platypus <a href="http://www.well.ox.ac.uk/platypus">http://www.well.ox.ac.uk/platypus</a><br>
</li>
<li>GATK <a href="https://www.broadinstitute.org/gatk/">https://www.broadinstitute.org/gatk/</a><br>
</li>
<li>ANNOVAR <a href="http://www.openbioinformatics.org/annovar/">http://www.openbioinformatics.org/annovar/</a><br>
</li>
</ul>

<p><strong>These tools require manual download and registration with the proivder. For non-academics/commercial groups, you will need to pay for some of these tools.</strong></p>

<h3>
<a id="dockerised-and-manual-builds" class="anchor" href="#dockerised-and-manual-builds" aria-hidden="true"><span class="octicon octicon-link"></span></a>Dockerised and Manual Builds</h3>

<table>
<thead>
<tr>
<th>Tool</th>
<th>Build</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_novoalign">novoalign</a></td>
<td>manual build</td>
</tr>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_annovar">annovar</a></td>
<td>manual build</td>
</tr>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_stampy">stampy</a></td>
<td>manual build</td>
</tr>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/nsgeasy_platypus">platypus</a></td>
<td>manual build</td>
</tr>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_gatk">gatk</a></td>
<td>manual build</td>
</tr>
</tbody>
</table>

<p>Once you have paid/registered and downloaded the tool, we provide scripts and guidance for building these tools on your system.  </p>

<p>Its as easy as:-  </p>

<div class="highlight highlight-bash"><pre>docker build -t compbio/ngseasy-<span class="pl-vo">${TOOL}</span> <span class="pl-s3">.</span></pre></div>

<hr>

<h3>
<a id="61-building-stampy" class="anchor" href="#61-building-stampy" aria-hidden="true"><span class="octicon octicon-link"></span></a>6.1 Building Stampy</h3>

<p><strong>resister at <a href="http://www.well.ox.ac.uk/project-stampy">http://www.well.ox.ac.uk/project-stampy</a></strong>  </p>

<p>Download stampy to local directory and check version number. If this differs from the <a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_stampy/Dockerfile">Dockerfile</a> build file, 
then edit the <a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_stampy/Dockerfile">Dockerfile</a> if needed. 
You will be emailed a URL to download stampy. Insert this into the <a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_stampy/Dockerfile">Dockerfile</a></p>

<div class="highlight highlight-bash"><pre><span class="pl-c"># on our local system we cd to media</span>
<span class="pl-s3">cd</span> /media

<span class="pl-c"># them move to ngs_projects toplevel directory</span>
<span class="pl-s3">cd</span> ngs_projects

<span class="pl-c"># and then the ngseasy folder with all our ngs scripts</span>
<span class="pl-c"># git  clone https://github.com/KHP-Informatics/ngs.git</span>
<span class="pl-c"># if you havent alreay</span>
<span class="pl-s3">cd</span> ngseasy

<span class="pl-c"># move to ngseasy_stampy folder</span>
<span class="pl-s3">cd</span> ngs/ngs_docker_debian/ngseasy_stampy

<span class="pl-c"># build</span>
docker build -t compbio/ngseasy-stampy:v1.0 <span class="pl-s3">.</span></pre></div>

<hr>

<h3>
<a id="62-building-platypus" class="anchor" href="#62-building-platypus" aria-hidden="true"><span class="octicon octicon-link"></span></a>6.2 Building Platypus</h3>

<p><strong>resister at <a href="http://www.well.ox.ac.uk/platypus">http://www.well.ox.ac.uk/platypus</a></strong></p>

<p>Download platypus to local directory and check version number. If this differs from the <a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/nsgeasy_platypus/Dockerfile">Dockerfile</a> build file, 
then edit the <a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/nsgeasy_platypus/Dockerfile">Dockerfile</a> if needed. 
You will be emailed a URL to download platypus. Insert this into the <a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/nsgeasy_platypus/Dockerfile">Dockerfile</a> </p>

<div class="highlight highlight-bash"><pre><span class="pl-c"># on our local system we cd to media</span>
<span class="pl-s3">cd</span> /media

<span class="pl-c"># them move to ngs_projects toplevel directory</span>
<span class="pl-s3">cd</span> ngs_projects

<span class="pl-c"># and then the ngseasy folder with all our ngs scripts</span>
<span class="pl-c"># git  clone https://github.com/KHP-Informatics/ngs.git</span>
<span class="pl-c"># if you havent alreay</span>
<span class="pl-s3">cd</span> ngseasy

<span class="pl-c"># move to ngseasy_stampy folder</span>
<span class="pl-s3">cd</span> ngs/ngs_docker_debian/ngseasy_platypus

<span class="pl-c"># build</span>
docker build -t compbio/ngseasy-platypus:v1.0 <span class="pl-s3">.</span></pre></div>

<hr>

<h3>
<a id="63-building-novoalign" class="anchor" href="#63-building-novoalign" aria-hidden="true"><span class="octicon octicon-link"></span></a>6.3 Building NOVOALIGN</h3>

<p><strong>Download Novoalign from  <a href="http://www.novocraft.com/">http://www.novocraft.com/</a></strong> into the local build directory <strong>ngs/ngs_docker_debian/ngseasy_novoalign</strong>. 
Edit the <a href="https://github.com/KHP-Informatics/ngs/blob/master/containerized/ngs_docker_debian/ngseasy_novoalign/Dockerfile">Dockerfile</a> to relfect
the correct version of novoalign.  </p>

<p>To use all novoalign fucntionality, you will need to <strong>pay for a license</strong>.   </p>

<p>Once you obtained your <strong>novoalign.lic</strong>, download this to the build directory <strong>ngs/ngs_docker_debian/ngseasy_novoalign</strong>, which now should contain your updated <a href="https://github.com/KHP-Informatics/ngs/blob/master/containerized/ngs_docker_debian/ngseasy_novoalign/Dockerfile">Dockerfile</a>.</p>

<div class="highlight highlight-bash"><pre><span class="pl-c"># on our local system we cd to media</span>
<span class="pl-s3">cd</span> /media

<span class="pl-c"># them move to ngs_projects toplevel directory</span>
<span class="pl-s3">cd</span> ngs_projects

<span class="pl-c"># and then the ngseasy folder with all our ngs scripts</span>
<span class="pl-c"># git  clone https://github.com/KHP-Informatics/ngs.git</span>
<span class="pl-c"># if you havent alreay</span>
<span class="pl-s3">cd</span> ngseasy

<span class="pl-c"># move to ngseasy_stampy folder</span>
<span class="pl-s3">cd</span> ngs/ngs_docker_debian/ngseasy_novoalign
ls </pre></div>

<p><strong>the directory should contain the following:-</strong></p>

<pre><code>Dockerfile
novoalign.lic
README.md
novosortV1.03.01.Linux3.0.tar.gz
novocraftV3.02.08.Linux3.0.tar.gz
</code></pre>

<p><strong>build novoalign</strong></p>

<div class="highlight highlight-bash"><pre><span class="pl-c"># build</span>
docker build -t compbio/ngseasy-novoalign:v1.0 <span class="pl-s3">.</span></pre></div>

<hr>

<h3>
<a id="64-building-gatk" class="anchor" href="#64-building-gatk" aria-hidden="true"><span class="octicon octicon-link"></span></a>6.4 Building GATK</h3>

<p>You need to register and accept the GATK license agreement at <a href="https://www.broadinstitute.org/gatk/">https://www.broadinstitute.org/gatk/</a>.  </p>

<p>Once done, download GATK and place in the GTAK build directory <strong>ngs/ngs_docker_debian/ngseasy_gatk</strong>.  </p>

<p>Edit the <a href="https://github.com/KHP-Informatics/ngs/blob/master/containerized/ngs_docker_debian/ngseasy_gatk/Dockerfile">Dockerfile</a> to relfect
the correct version of GATK.  </p>

<div class="highlight highlight-bash"><pre><span class="pl-c"># on our local system we cd to media</span>
<span class="pl-s3">cd</span> /media

<span class="pl-c"># them move to ngs_projects toplevel directory</span>
<span class="pl-s3">cd</span> ngs_projects

<span class="pl-c"># and then the ngseasy folder with all our ngs scripts</span>
<span class="pl-c"># git  clone https://github.com/KHP-Informatics/ngs.git</span>
<span class="pl-c"># if you havent alreay</span>
<span class="pl-s3">cd</span> ngseasy

<span class="pl-c"># move to ngseasy_stampy folder</span>
<span class="pl-s3">cd</span> ngs/ngs_docker_debian/ngseasy_gatk
ls </pre></div>

<p><strong>the directory should contain the following:-</strong></p>

<pre><code>Dockerfile
README.md
GenomeAnalysisTK-3.3-0.tar.bz2
</code></pre>

<p><strong>build gatk</strong></p>

<div class="highlight highlight-bash"><pre><span class="pl-c"># build</span>
docker build -t compbio/ngseasy-gatk:v1.0 <span class="pl-s3">.</span></pre></div>

<hr>

<h2>
<a id="7-manually-build-ngseasy-variant-annotaion-container-images" class="anchor" href="#7-manually-build-ngseasy-variant-annotaion-container-images" aria-hidden="true"><span class="octicon octicon-link"></span></a>7. Manually Build NGSeasy Variant Annotaion Container Images</h2>

<p>The tools used for variant annotation use large databases and the docker images exceed 10GB. Therefore, the user should manually build these container images prior to running the NGS pipelines.
Docker build files (<a href="https://docs.docker.com/jsearch/?q=Dockerfile">Dockerfile</a>) are available for </p>

<ul>
<li>
<a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_annovar/Dockerfile">Annovar</a><br>
</li>
<li>
<a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_vep/Dockerfile">VEP</a><br>
</li>
<li>
<a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_snpeff/Dockerfile">snpEff</a><br>
</li>
</ul>

<p><strong>Note</strong> Annovar requires user registration.  </p>

<p><strong>Once built on the user system, these container images can persist for as long as the user wants.</strong>  </p>

<p><strong>Large Variant Annotation Container Images</strong></p>

<table>
<thead>
<tr>
<th>Tool</th>
<th>Build</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_annovar">annovar</a></td>
<td>manual build</td>
</tr>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_vep">vep</a></td>
<td>manual build</td>
</tr>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/tree/master/containerized/ngs_docker_debian/ngseasy_snpeff">snpeff</a></td>
<td>manual build</td>
</tr>
</tbody>
</table>

<p>Its as easy as:-  </p>

<div class="highlight highlight-bash"><pre>docker build -t compbio/ngseasy-<span class="pl-vo">${TOOL}</span> <span class="pl-s3">.</span></pre></div>

<hr>

<h3>
<a id="71-build-vep" class="anchor" href="#71-build-vep" aria-hidden="true"><span class="octicon octicon-link"></span></a>7.1 Build VEP</h3>

<div class="highlight highlight-bash"><pre>
<span class="pl-s3">cd</span> /media/ngs_projects/nsgeasy/ngs/containerized/ngs_docker_debian/ngseasy_vep

sudo docker build -t compbio/ngseasy-vep:<span class="pl-vo">${VERSION}</span> <span class="pl-s3">.</span></pre></div>

<hr>

<h3>
<a id="72-build-annovar" class="anchor" href="#72-build-annovar" aria-hidden="true"><span class="octicon octicon-link"></span></a>7.2 Build Annovar</h3>

<div class="highlight highlight-bash"><pre><span class="pl-s3">cd</span> /media/ngs_projects/nsgeasy/ngs/containerized/ngs_docker_debian/ngseasy_annovar

sudo docker build -t compbio/ngseasy-annovar:<span class="pl-vo">${VERSION}</span> <span class="pl-s3">.</span></pre></div>

<hr>

<h3>
<a id="73-build-snpeff" class="anchor" href="#73-build-snpeff" aria-hidden="true"><span class="octicon octicon-link"></span></a>7.3 Build snpEff</h3>

<div class="highlight highlight-bash"><pre><span class="pl-s3">cd</span> /media/ngs_projects/nsgeasy/ngs/containerized/ngs_docker_debian/ngseasy_snpeff

sudo docker build -t compbio/ngseasy-snpeff:<span class="pl-vo">${VERSION}</span> <span class="pl-s3">.</span></pre></div>

<hr>

<h2>
<a id="8-set-up-ngseasy-project-working-directories" class="anchor" href="#8-set-up-ngseasy-project-working-directories" aria-hidden="true"><span class="octicon octicon-link"></span></a>8. Set up NGSeasy Project Working Directories</h2>

<p>Running the script <code>ngseasy_initiate_project</code> ensures that all relevant directories are set up, and also enforces a clean structure to the NGS project.  </p>

<p>Within this we make a <code>raw_fastq</code> folder, where we temporarily store all the raw fastq files for each project. 
This folder acts as an initial stagging area for the raw fastq files. During the project set up, we copy/move project/sample related fastq files to their own specific directories.
Fastq files must have suffix and be gzipped: <strong>_1.fq.gz</strong> or <strong>_2.fq.gz</strong><br>
furture version will allow any format  </p>

<p>Running <code>ngseasy_initiate_project</code> with the relevent configuration file, will set up the following directory structure for every project and sample within a project:-  </p>

<h2>
<a id="ngs-project-directory" class="anchor" href="#ngs-project-directory" aria-hidden="true"><span class="octicon octicon-link"></span></a>NGS Project Directory</h2>

<div class="highlight highlight-bash"><pre><span class="pl-s3">.</span>
ngs_projects  
<span class="pl-k">|</span>  
<span class="pl-k">|</span>__raw_fastq  
<span class="pl-k">|</span>__config_files  
<span class="pl-k">|</span>__reference_genomes_b37  
<span class="pl-k">|</span>__gatk_resources  
<span class="pl-k">|</span>__ngseasy
<span class="pl-k">|</span>
<span class="pl-k">|</span>__ project_id  
    <span class="pl-k">|</span>  
    <span class="pl-k">|</span>__run_logs  
    <span class="pl-k">|</span>__config_files  
    <span class="pl-k">|</span>__project_vcfs  
    <span class="pl-k">|</span>__project_bams  
    <span class="pl-k">|</span>__project_reports  
    <span class="pl-k">|</span>
    <span class="pl-k">|</span>__sample_id_1  
    <span class="pl-k">|</span>   <span class="pl-k">|</span>  
    <span class="pl-k">|</span>   <span class="pl-k">|</span>__fastq  
    <span class="pl-k">|</span>   <span class="pl-k">|</span>__tmp  
    <span class="pl-k">|</span>   <span class="pl-k">|</span>__alignments  
    <span class="pl-k">|</span>   <span class="pl-k">|</span>__vcf  
    <span class="pl-k">|</span>   <span class="pl-k">|</span>__reports  
    <span class="pl-k">|</span>   <span class="pl-k">|</span>__config_files  
    <span class="pl-k">|</span>
    <span class="pl-k">|</span>
    <span class="pl-k">|</span>__sample_id_n  
        <span class="pl-k">|</span>  
        <span class="pl-k">|</span>__fastq  
        <span class="pl-k">|</span>__tmp  
        <span class="pl-k">|</span>__alignments  
        <span class="pl-k">|</span>__vcf  
        <span class="pl-k">|</span>__reports  
        <span class="pl-k">|</span>__config_files  </pre></div>

<p><strong>Running <code>ngseasy_initiate_project</code></strong></p>

<div class="highlight highlight-bash"><pre>ngseasy_initiate_project -c config.file.tsv -d /media/ngs_projects</pre></div>

<hr>

<h2>
<a id="9-ngseasy-project-configuration-file" class="anchor" href="#9-ngseasy-project-configuration-file" aria-hidden="true"><span class="octicon octicon-link"></span></a>9. NGSeasy Project configuration file</h2>

<p>In Excel make config file and save as [TAB] Delimited file with <code>.tsv</code> extenstion.<br>
See Example provided and <a href="https://docs.google.com/spreadsheets/d/1kp1Nyw0x3zXqO2Wm2Z25ErJ0Z-Uoab8tjRPq9h4sonk/edit?usp=sharing">GoogleDoc</a>. Remove the header from this file before running the pipeline. This sets up Information related to: Project Name, Sample Name, Library Type, Pipeline to call, NCPU.</p>

<p>The [config.file.tsv] should contain the following 15 columns for each sample to be run through a pipeline:- </p>

<table>
<thead>
<tr>
<th>Variable</th>
<th>type</th>
<th>Description</th>
<th>Options/Examples</th>
</tr>
</thead>
<tbody>
<tr>
<td>POJECT_ID</td>
<td>string</td>
<td>Project ID</td>
<td>Cancer</td>
</tr>
<tr>
<td>SAMPLE_ID</td>
<td>string</td>
<td>Sample ID</td>
<td>T100</td>
</tr>
<tr>
<td>FASTQ1</td>
<td>string</td>
<td>Raw fastq file name read 1</td>
<td>foo_1_fq.gz</td>
</tr>
<tr>
<td>FASTQ2</td>
<td>string</td>
<td>Raw fastq file name read 1</td>
<td>foo_2_fq.gz</td>
</tr>
<tr>
<td>PROJECT_DIR</td>
<td>string</td>
<td>Project Directory</td>
<td>/medida/ngs_projects</td>
</tr>
<tr>
<td>DNA_PREP_LIBRARY_ID</td>
<td>string</td>
<td>DNA Libray Prep ID</td>
<td>Custom_Cancer</td>
</tr>
<tr>
<td>NGS_PLATFORM</td>
<td>string</td>
<td>Platform Name</td>
<td>ILLUMINA</td>
</tr>
<tr>
<td>NGS_TYPE</td>
<td>string</td>
<td>Experiment type</td>
<td>WGS/WEX/TGS/</td>
</tr>
<tr>
<td>BED_ANNO</td>
<td>string</td>
<td>Annotation Bed File</td>
<td>exons_b37.bed</td>
</tr>
<tr>
<td>PIPELINE</td>
<td>string</td>
<td>NGSeasy Pipeline Script</td>
<td>ngs_full_gatk/ngs_full_no_gatk</td>
</tr>
<tr>
<td>ALIGNER</td>
<td>string</td>
<td>Aligner</td>
<td>bwa/bowtie/stampy/novoalign</td>
</tr>
<tr>
<td>VARCALLER</td>
<td>string</td>
<td>Variant Caller</td>
<td>ensemble/freebayes/platypus/UnifiedGenotyper/HaplotypeCaller</td>
</tr>
<tr>
<td>GTMODEGATK</td>
<td>string</td>
<td>GATK Variant Caller Mode</td>
<td>EMIT_ALL_CONFIDENT_SITES/EMIT_VARIANTS_ONLY</td>
</tr>
<tr>
<td>CLEANUP</td>
<td>string</td>
<td>Clean Up Files (TRUE/FALSE)</td>
<td>TRUE/FALSE</td>
</tr>
<tr>
<td>NCPU</td>
<td>number</td>
<td>Number of cores to call</td>
<td>1..n</td>
</tr>
<tr>
<td>VERSION</td>
<td>number</td>
<td>NGSeasy Version</td>
<td>v0.9/v1.0</td>
</tr>
</tbody>
</table>

<p>In the config file we set PIPELINE to call the pipeline <strong>[ngs_full_gatk]</strong> or <strong>[ngs_full_no_gatk]</strong>.   </p>

<p><em>coming soon</em> options to add user email, specify non-gatk runs  </p>

<hr>

<h2>
<a id="10-copy-project-fastq-files-to-relevent-projectsample-directories" class="anchor" href="#10-copy-project-fastq-files-to-relevent-projectsample-directories" aria-hidden="true"><span class="octicon octicon-link"></span></a>10. Copy Project Fastq files to relevent Project/Sample Directories</h2>

<div class="highlight highlight-bash"><pre>ngseasy_initiate_fastq -c config.file.tsv -d /media/ngs_projects</pre></div>

<hr>

<h2>
<a id="11-start-the-ngseasy-volume-contaier" class="anchor" href="#11-start-the-ngseasy-volume-contaier" aria-hidden="true"><span class="octicon octicon-link"></span></a>11. Start the NGSeasy Volume Contaier</h2>

<p>In the Docker container the project directory is mounted in <code>/home/pipeman/ngs_projects</code></p>

<div class="highlight highlight-bash"><pre>ngseasy_volumes_container -d /media/ngs_projects</pre></div>

<p><strong>inside ngseasy_volumes_container</strong>. This is what it is calling. Note the directory names and mounts. </p>

<div class="highlight highlight-bash"><pre>
<span class="pl-c"># host_vol_dir = ngs_projects</span>

  docker run \
  -d \
  -P \
  -v <span class="pl-vo">${host_vol_dir}</span>/fastq_raw:/home/pipeman/fastq_raw \
  -v <span class="pl-vo">${host_vol_dir}</span>/reference_genomes_b37:/home/pipeman/reference_genomes_b37 \
  -v <span class="pl-vo">${host_vol_dir}</span>/gatk_resources:/home/pipeman/gatk_resources \
  -v <span class="pl-vo">${host_vol_dir}</span>:/home/pipeman/ngs_projects \
  -v <span class="pl-vo">${host_vol_dir}</span>/ngseasy/ngs/bin:/home/pipeman/ngseasy_scripts \
  --name volumes_container \
  -t compbio/ngseasy-base:wheezy</pre></div>

<hr>

<h2>
<a id="12-running-an-ngseasy-full-pipeline--from-raw-fastq-to-vcf-calls" class="anchor" href="#12-running-an-ngseasy-full-pipeline--from-raw-fastq-to-vcf-calls" aria-hidden="true"><span class="octicon octicon-link"></span></a>12. Running an NGSeasy full pipeline : from raw fastq to vcf calls</h2>

<p><strong>run ngseay</strong></p>

<div class="highlight highlight-bash"><pre>
    ngseasy -c config.file.tsv -d /media/nsg_projects
</pre></div>

<p>The pipeline is defined in the config file as <strong>[ngs_full_gatk]</strong></p>

<hr>

<h2>
<a id="the-ngseasy-pipelines" class="anchor" href="#the-ngseasy-pipelines" aria-hidden="true"><span class="octicon octicon-link"></span></a>The NGSeasy Pipelines</h2>

<table>
<thead>
<tr>
<th>Pipeline</th>
<th>Short Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/blob/master/bin/ngs_full_gatk">ngs_full_gatk</a></td>
<td>fastq to recalibrated bam to vcf using GATK</td>
</tr>
<tr>
<td><a href="https://github.com/KHP-Informatics/ngs/blob/master/bin/ngs_full_no_gatk">ngs_full_no_gatk</a></td>
<td>fastq to recalibrated bam to vcf</td>
</tr>
</tbody>
</table>

<p>gatk version includes indel realignment and base recalibration.  </p>

<p>Non-academics/commercial groups need to pay for GATK.  </p>

<p>Currently <strong>ngs_full_gatk</strong> pipeline is the most developed module.  </p>

<p>The <strong>ngs_full_no_gatk</strong> pipeline provides alternatives to processing with GATK. Here BamUtil:recab is used to recalibrate base quality scores and freebayes/platypus are the variant callers of choice.</p>

<h2>
<a id="ngs_full_gatk" class="anchor" href="#ngs_full_gatk" aria-hidden="true"><span class="octicon octicon-link"></span></a>ngs_full_gatk</h2>

<p>Each <strong>pipeline</strong> is a bash wrapper that calls a number of functions/steps set out in <a href="https://github.com/KHP-Informatics/ngs#the-full-ngseasy-pipeline">The Full NGSeasy pipeline</a>.</p>

<p>Here  <strong>[ngs_full_gatk]</strong> is a wrapper/fucntion for calling an NGS pipeline. The inside to this script is set out below:-  </p>

<div class="highlight highlight-bash"><pre><span class="pl-c">#!/bin/bash -x</span>

<span class="pl-c">#usage printing func</span>
<span class="pl-en">usage</span>()
{
cat <span class="pl-k">&lt;&lt;</span> EOF
  This script calls the NGSeasy pipeline ngs_full_gatk

  ARGUMENTS:
  -h      Flag: Show this <span class="pl-s3">help</span> message
  -c      NGSeasy project and run configureation file
  -d      NGSeasy project directory

  EXAMPLE USAGE:

    ngseasy -c config.file.tsv -d project_directory

EOF
}

<span class="pl-c">#get options for command line args</span>
  <span class="pl-k">while</span>  <span class="pl-s3">getopts</span> <span class="pl-s1"><span class="pl-pds">"</span>hc:d:<span class="pl-pds">"</span></span> opt
  <span class="pl-k">do</span>

      <span class="pl-k">case</span> <span class="pl-vo">${opt}</span> in
      h)
      usage <span class="pl-c">#print help</span>
      <span class="pl-s3">exit</span> 0
      ;;

      c)
      config_tsv=<span class="pl-vo">${OPTARG}</span>
      ;;

      d)
      project_directory=<span class="pl-vo">${OPTARG}</span>
      ;; 
      <span class="pl-k">esac</span>
  <span class="pl-k">done</span>

<span class="pl-c">#check config file exists.</span>
<span class="pl-k">if</span> [ <span class="pl-k">!</span> -e <span class="pl-s1"><span class="pl-pds">"</span><span class="pl-vo">${config_tsv}</span><span class="pl-pds">"</span></span> ] 
<span class="pl-k">then</span>
        <span class="pl-s3">echo</span> <span class="pl-s1"><span class="pl-pds">"</span>ERROR :  <span class="pl-vo">${config_tsv}</span> does not exist <span class="pl-pds">"</span></span>
        usage<span class="pl-k">;</span>
        <span class="pl-s3">exit</span> 1<span class="pl-k">;</span>
<span class="pl-k">fi</span>

<span class="pl-c">#check exists.</span>
  <span class="pl-k">if</span> [ <span class="pl-k">!</span> -d <span class="pl-s1"><span class="pl-pds">"</span><span class="pl-vo">${project_directory}</span><span class="pl-pds">"</span></span> ] 
  <span class="pl-k">then</span>
      <span class="pl-s3">echo</span> <span class="pl-s1"><span class="pl-pds">"</span> ERROR : <span class="pl-vo">${project_directory}</span> does not exist <span class="pl-pds">"</span></span>
      usage<span class="pl-k">;</span>
      <span class="pl-s3">exit</span> 1<span class="pl-k">;</span>
  <span class="pl-k">fi</span>

<span class="pl-c">##################  </span>
<span class="pl-c"># start pipeline #</span>
<span class="pl-c">##################</span>

<span class="pl-c"># Each of these fucntions will call the required image/container(s) and run a part of the NGS pipeline. Each step is usually </span>
<span class="pl-c"># dependent on the previous step(s) - in that they require certain data/input/output in the correct format </span>
<span class="pl-c"># and with the correct nameing conventions enforced by our pipeline to exist, before executing.</span>

ngseasy_fastqc  -c <span class="pl-vo">${config_tsv}</span> -d <span class="pl-vo">${project_directory}</span>

ngseasy_trimmomatic -c <span class="pl-vo">${config_tsv}</span> -d <span class="pl-vo">${project_directory}</span>

ngseasy_alignment -c <span class="pl-vo">${config_tsv}</span> -d <span class="pl-vo">${project_directory}</span>

ngseasy_addreadgroup -c <span class="pl-vo">${config_tsv}</span> -d <span class="pl-vo">${project_directory}</span>

ngseasy_markduplicates -c <span class="pl-vo">${config_tsv}</span> -d <span class="pl-vo">${project_directory}</span>

ngseasy_indel_realn -c <span class="pl-vo">${config_tsv}</span> -d <span class="pl-vo">${project_directory}</span>

ngseasy_base_recal -c <span class="pl-vo">${config_tsv}</span> -d <span class="pl-vo">${project_directory}</span>

ngseasy_filter_recalbam -c <span class="pl-vo">${config_tsv}</span> -d <span class="pl-vo">${project_directory}</span>

ngseasy_alignment_qc -c <span class="pl-vo">${config_tsv}</span> -d <span class="pl-vo">${project_directory}</span>

ngseasy_variant_calling -c <span class="pl-vo">${config_tsv}</span> -d <span class="pl-vo">${project_directory}</span>

<span class="pl-c"># coming soon...</span>
<span class="pl-c"># ngseasy_filter_bam -c ${config_tsv} -d ${project_directory}</span>
<span class="pl-c"># ngseasy_cnv_calling -c ${config_tsv} -d ${project_directory}</span>
<span class="pl-c"># ngseasy_variant_filtering -c ${config_tsv} -d ${project_directory}</span>
<span class="pl-c"># ngseasy_variant_annotation -c ${config_tsv} -d ${project_directory}</span>
<span class="pl-c"># ngseasy_report -c ${config_tsv} -d ${project_directory}</span></pre></div>

<hr>

<h1>
<a id="output-suffixes-" class="anchor" href="#output-suffixes-" aria-hidden="true"><span class="octicon octicon-link"></span></a>Output suffixes </h1>

<h3>
<a id="alignment-output" class="anchor" href="#alignment-output" aria-hidden="true"><span class="octicon octicon-link"></span></a>Alignment Output</h3>

<p>*.raw.sam  (WEX ~ 8GB)
*.raw.bam<br>
*.raw.bai<br>
*.sort.bam  (WEX ~ 3GB)
*.sort.bai  </p>

<hr>

<h3>
<a id="addreadgroup" class="anchor" href="#addreadgroup" aria-hidden="true"><span class="octicon octicon-link"></span></a>Addreadgroup</h3>

<p>*.addrg.bam<br>
*.addrg.bai<br>
*.addrg.bam.bai  </p>

<hr>

<h3>
<a id="dupemark" class="anchor" href="#dupemark" aria-hidden="true"><span class="octicon octicon-link"></span></a>Dupemark</h3>

<p>*.dupemk.bam<br>
*.dupemk.bai
*.dupemk.bam.bai  </p>

<hr>

<h3>
<a id="indel-realign" class="anchor" href="#indel-realign" aria-hidden="true"><span class="octicon octicon-link"></span></a>Indel realign</h3>

<p>*.realn.bam<br>
*.realn.bai
*.realn.bam.bai  </p>

<hr>

<h3>
<a id="base-recal" class="anchor" href="#base-recal" aria-hidden="true"><span class="octicon octicon-link"></span></a>Base recal</h3>

<p>*.recal.bam  (WEX ~ 4.4G)<br>
*.recal.bai<br>
*.recal.bam.bai<br>
*.realn.bam.BaseRecalibrator.table<br>
*.recal.bam.BaseRecalibrator.table<br>
*.recal.bam.BaseRecalibrator.BQSR.csv  </p>

<hr>

<h2>
<a id="thresholds-for-variant-calling-etc" class="anchor" href="#thresholds-for-variant-calling-etc" aria-hidden="true"><span class="octicon octicon-link"></span></a>Thresholds for Variant calling etc</h2>

<p>For Freebayes and Platypus tools:-  </p>

<ul>
<li>We set min coverage to 10<br>
</li>
<li>Min mappinng quality to 20<br>
</li>
<li>Min base quality to 20</li>
</ul>

<p>For GATK HaplotypeCaller (and UnifiedGenotyper)</p>

<p><code>-stand_call_conf 30 -stand_emit_conf 10 -dcov 250 -minPruning 10</code></p>

<p>Note: <code>minPruning 10</code> was added as many runs of HaplotypeCaller failed when using non-bwa aligend and GATK best practices cleaned BAMs. This fix sorted all problems out, and you really dont want dodgy variant calls...do you? Same goes for thresholds hard coded for use with Freebayes and Platypus.<br>
These setting all work well in our hands. Feel  free to edit the scripts to suit your needs.</p>

<hr>

<p>blah blah blah</p>

<hr>

<h2>
<a id="gottchas" class="anchor" href="#gottchas" aria-hidden="true"><span class="octicon octicon-link"></span></a>Gottchas</h2>

<p><strong>bin/bash -c</strong></p>

<ul>
<li>need to add <code>/bin/bash -c ${COMMAND}</code> when software require <code>&gt;</code> redirect to some output</li>
</ul>

<p>example below for bwa:-  </p>

<pre><code>  sudo docker run \
  -P \
  --name sam2bam_${SAMPLE_ID} \
  --volumes-from volumes_container \
  -t compbio/ngseasy-samtools:v0.9 /bin/bash -c \
  "/usr/local/pipeline/samtools/samtools view -bhS ${SOUTDocker}/alignments/${BAM_PREFIX}.raw.bwa.sam &gt; ${SOUTDocker}/alignments/${BAM_PREFIX}.raw.bwa.bam"
</code></pre>

<p>runnig this without <code>/bin/bash -c</code> breaks. The <code>&gt;</code> is called outside of the container</p>

<h3>
<a id="the-annoying-thing-about-gatk" class="anchor" href="#the-annoying-thing-about-gatk" aria-hidden="true"><span class="octicon octicon-link"></span></a>The Annoying thing about GATK!</h3>

<p>This will break your runs if multiple calls try and access the file when the first call deletes it!  </p>

<pre><code>WARN  11:05:27,577 RMDTrackBuilder - Index file /home/pipeman/gatk_resources/Mills_and_1000G_gold_standard.indels.b37.vcf.idx is out of date (index older than input file), deleting and updating the index file 
INFO  11:05:31,699 RMDTrackBuilder - Writing Tribble index to disk for file /home/pipeman/gatk_resources/Mills_and_1000G_gold_standard.indels.b37.vcf.idx 
</code></pre>

<h2>
<a id="cnv-tools-to-think-about" class="anchor" href="#cnv-tools-to-think-about" aria-hidden="true"><span class="octicon octicon-link"></span></a>CNV tools to think about</h2>

<p>EXCAVATOR: detecting copy number variants from whole-exome sequencing data @ <a href="http://genomebiology.com/2013/14/10/R120">http://genomebiology.com/2013/14/10/R120</a></p>

<blockquote>
<p>We developed a novel software tool, EXCAVATOR, for the detection of copy number variants (CNVs) from whole-exome sequencing data. EXCAVATOR combines a three-step normalization procedure with a novel heterogeneous hidden Markov model algorithm and a calling method that classifies genomic regions into five copy number states. We validate EXCAVATOR on three datasets and compare the results with three other methods. These analyses show that EXCAVATOR outperforms the other methods and is therefore a valuable tool for the investigation of CNVs in largescale projects, as well as in clinical research and diagnostics. EXCAVATOR is freely available at <a href="http://sourceforge.net/projects/excavatortool/">http://sourceforge.net/projects/excavatortool/</a> webcite.</p>
</blockquote>

<hr>

<h2>
<a id="useful-links" class="anchor" href="#useful-links" aria-hidden="true"><span class="octicon octicon-link"></span></a>Useful Links</h2>

<ul>
<li>
<a href="https://bcbio.wordpress.com/">https://bcbio.wordpress.com/</a><br>
</li>
<li>
<a href="https://basecallbio.wordpress.com/2013/04/23/base-quality-score-rebinning/">https://basecallbio.wordpress.com/2013/04/23/base-quality-score-rebinning/</a><br>
</li>
<li>
<a href="https://github.com/statgen/bamUtil">https://github.com/statgen/bamUtil</a><br>
</li>
<li>
<a href="http://genome.sph.umich.edu/wiki/BamUtil:_recab">http://genome.sph.umich.edu/wiki/BamUtil:_recab</a><br>
</li>
<li>
<a href="https://github.com/chapmanb/bcbio.variation">https://github.com/chapmanb/bcbio.variation</a><br>
</li>
<li><a href="http://plagnol-lab.blogspot.co.uk/2013/11/faq-and-clarifications-for-exomedepth-r.html">http://plagnol-lab.blogspot.co.uk/2013/11/faq-and-clarifications-for-exomedepth-r.html</a></li>
</ul>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Ngseasy maintained by <a href="https://github.com/KHP-Informatics">KHP-Informatics</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
